diff -urN mc-4.7.5.5.orig/lib/fileloc.h mc-4.7.5.5/lib/fileloc.h
--- mc-4.7.5.5.orig/lib/fileloc.h	2011-09-28 12:48:35.000000000 -0400
+++ mc-4.7.5.5/lib/fileloc.h	2011-10-06 18:13:15.000000000 -0400
@@ -38,6 +38,7 @@
 #define FISH_UNLINK_FILE        "unlink"
 #define FISH_CHOWN_FILE         "chown"
 #define FISH_CHMOD_FILE         "chmod"
+#define FISH_UTIME_FILE         "utime"
 #define FISH_RMDIR_FILE         "rmdir"
 #define FISH_LN_FILE            "ln"
 #define FISH_MV_FILE            "mv"
diff -urN mc-4.7.5.5.orig/lib/vfs/mc-vfs/fish.c mc-4.7.5.5/lib/vfs/mc-vfs/fish.c
--- mc-4.7.5.5.orig/lib/vfs/mc-vfs/fish.c	2011-09-28 12:48:35.000000000 -0400
+++ mc-4.7.5.5/lib/vfs/mc-vfs/fish.c	2011-10-06 18:15:20.000000000 -0400
@@ -275,6 +275,7 @@
     g_free (SUP.scr_unlink);
     g_free (SUP.scr_chown);
     g_free (SUP.scr_chmod);
+    g_free (SUP.scr_utime);
     g_free (SUP.scr_rmdir);
     g_free (SUP.scr_ln);
     g_free (SUP.scr_mv);
@@ -575,6 +576,7 @@
     SUP.scr_unlink = fish_load_script_from_file (host, FISH_UNLINK_FILE, FISH_UNLINK_DEF_CONTENT);
     SUP.scr_chown = fish_load_script_from_file (host, FISH_CHOWN_FILE, FISH_CHOWN_DEF_CONTENT);
     SUP.scr_chmod = fish_load_script_from_file (host, FISH_CHMOD_FILE, FISH_CHMOD_DEF_CONTENT);
+    SUP.scr_utime = fish_load_script_from_file (host, FISH_UTIME_FILE, FISH_UTIME_DEF_CONTENT);
     SUP.scr_rmdir = fish_load_script_from_file (host, FISH_RMDIR_FILE, FISH_RMDIR_DEF_CONTENT);
     SUP.scr_ln = fish_load_script_from_file (host, FISH_LN_FILE, FISH_LN_DEF_CONTENT);
     SUP.scr_mv = fish_load_script_from_file (host, FISH_MV_FILE, FISH_MV_DEF_CONTENT);
@@ -1272,6 +1274,35 @@
 /* --------------------------------------------------------------------------------------------- */
 
 static int
+fish_utime (struct vfs_class *me, const char *path, struct utimbuf *times)
+{
+    gchar *shell_commands = NULL;
+    char utcmtime[16], utcatime[16];
+    struct tm *gmt;
+
+    PREFIX;
+
+    gmt = gmtime(&times->modtime);
+    g_snprintf (utcmtime, sizeof (utcmtime), "%04d%02d%02d%02d%02d.%02d",
+                gmt->tm_year+1900, gmt->tm_mon+1, gmt->tm_mday,
+                gmt->tm_hour, gmt->tm_min, gmt->tm_sec);
+
+    gmt = gmtime(&times->actime);
+    g_snprintf (utcatime, sizeof (utcatime), "%04d%02d%02d%02d%02d.%02d",
+                gmt->tm_year+1900, gmt->tm_mon+1, gmt->tm_mday,
+                gmt->tm_hour, gmt->tm_min, gmt->tm_sec);
+
+    shell_commands = g_strconcat (SUP.scr_env, "FISH_FILENAME=%s FISH_FILEMTIME=%s FISH_FILEATIME=%s;\n",
+                                  SUP.scr_utime, (char *) NULL);
+    g_snprintf (buf, sizeof (buf), shell_commands, rpath, utcmtime, utcatime);
+    g_free (shell_commands);
+    g_free (rpath);
+    return fish_send_command (me, super, buf, OPT_FLUSH);
+}
+
+/* --------------------------------------------------------------------------------------------- */
+
+static int
 fish_unlink (struct vfs_class *me, const char *path)
 {
     gchar *shell_commands = NULL;
@@ -1458,6 +1489,7 @@
     vfs_fish_ops.fill_names = fish_fill_names;
     vfs_fish_ops.chmod = fish_chmod;
     vfs_fish_ops.chown = fish_chown;
+    vfs_fish_ops.utime = fish_utime;
     vfs_fish_ops.open = fish_open;
     vfs_fish_ops.symlink = fish_symlink;
     vfs_fish_ops.link = fish_link;
diff -urN mc-4.7.5.5.orig/lib/vfs/mc-vfs/fishdef.h mc-4.7.5.5/lib/vfs/mc-vfs/fishdef.h
--- mc-4.7.5.5.orig/lib/vfs/mc-vfs/fishdef.h	2011-09-28 12:48:35.000000000 -0400
+++ mc-4.7.5.5/lib/vfs/mc-vfs/fishdef.h	2011-10-06 18:10:44.000000000 -0400
@@ -74,6 +74,19 @@
 "    echo \"### 500\"\n"                                                    \
 "fi\n"
 
+/* default 'utime' script */
+#define FISH_UTIME_DEF_CONTENT ""                                                     \
+"#UTIME $FISH_FILEMTIME $FISH_FILEATIME $FISH_FILENAME\n"                             \
+"if TZ=UTC touch -m -t $FISH_FILEMTIME \"/${FISH_FILENAME}\" 2>/dev/null; then\n"     \
+"    if TZ=UTC touch -a -t $FISH_FILEATIME \"/${FISH_FILENAME}\" 2>/dev/null; then\n" \
+"        echo \"### 000\"\n"                                                          \
+"    else\n"                                                                          \
+"        echo \"### 500\"\n"                                                          \
+"    fi\n"                                                                            \
+"else\n"                                                                              \
+"    echo \"### 500\"\n"                                                              \
+"fi\n"
+
 /* default 'rmdir' script */
 #define FISH_RMDIR_DEF_CONTENT ""                                           \
 "#RMD $FISH_FILENAME\n"                                                     \
diff -urN mc-4.7.5.5.orig/lib/vfs/mc-vfs/xdirentry.h mc-4.7.5.5/lib/vfs/mc-vfs/xdirentry.h
--- mc-4.7.5.5.orig/lib/vfs/mc-vfs/xdirentry.h	2011-09-28 12:49:09.000000000 -0400
+++ mc-4.7.5.5/lib/vfs/mc-vfs/xdirentry.h	2011-10-06 18:15:09.000000000 -0400
@@ -72,6 +72,7 @@
             int flags;
             char *scr_ls;
             char *scr_chmod;
+            char *scr_utime;
             char *scr_exists;
             char *scr_mkdir;
             char *scr_unlink;
