diff --git a/src/execute.c b/src/execute.c
index 999ae0e..05086ce 100644
--- a/src/execute.c
+++ b/src/execute.c
@@ -549,16 +549,16 @@ toggle_panels (void)
     vfs_path_free (new_dir_vpath);
 #endif /* ENABLE_SUBSHELL */
 
+    if (was_sigwinch != 0 || mc_global.tty.winch_flag != 0)
+        dialog_change_screen_size ();
+    else
+        repaint_screen ();
+
     if (mc_global.mc_run_mode == MC_RUN_FULL)
     {
         update_panels (UP_OPTIMIZE, UP_KEEPSEL);
         update_xterm_title_path ();
     }
-
-    if (was_sigwinch != 0 || mc_global.tty.winch_flag != 0)
-        dialog_change_screen_size ();
-    else
-        repaint_screen ();
 }
 
 /* --------------------------------------------------------------------------------------------- */
