From 22fad061c612abb08c56e1d1c6ebd71d1cb59b63 Mon Sep 17 00:00:00 2001
From: Sebastian Gniazdowski <sgniazdowski@gmail.com>
Date: Mon, 8 Feb 2021 23:08:35 -0600
Subject: New utility make targets for tags.

---
 Makefile.am | 49 ++++++++++++++++++++++++++++++++++++++++++++++++-
 1 file changed, 48 insertions(+), 1 deletion(-)

diff --git a/Makefile.am b/Makefile.am
index f86f6ed38..196a474af 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -25,8 +25,55 @@ CONFIG_STATUS_DEPENDENCIES = $(top_srcdir)/version.h
         cppcheck-portability \
         cppcheck-style \
         cppcheck-warning \
-        cppcheck-all
+        cppcheck-all \
+        tags \
+        tags-emacs \
+        tags-vim
 
+# Invoke a ctags utility to generate Emacs and Vim's format tag file.
+# The meaning of the flags:
+# - `p` kind enables function prototypes,
+# - `x` kind enables extern and forward variable declarations,
+# - `S` field enables function signatures,
+# - `properties` field enables static, volatile, etc. properties,
+# - `macrodef` field enables macro definitions.
+#
+# Generate two formats at once: Emacs and Vim's.
+tags: tags-emacs tags-vim
+
+# Build only the Emacs-style `TAGS` file.
+tags-emacs:
+	@if type ctags >/dev/null 2>&1; then \
+		if ctags --version | grep >/dev/null 2>&1 "Universal Ctags"; then \
+			ctags -e -R --languages=C --langmap=C:.h.c --c-kinds=+px \
+			    --fields=+S --fields-c=+"{properties}" --fields-c=+"{macrodef}" \
+			    --extras=+qr; \
+		else \
+			ctags -e -R --languages=C --langmap=C:.h.c --c-kinds=+px \
+			    --fields=+S; \
+		fi; \
+		printf "Created the Emacs \`TAGS\` file.\\n"; \
+	else \
+	    printf 'Error: Please install a Ctags (e.g.: either the Exuberant or Universal %b' \
+	    'version) utility first.\n'; \
+	fi
+
+# Build only the Vim-style `tags` file.
+tags-vim:
+	@if type ctags >/dev/null 2>&1; then \
+		if ctags --version | grep >/dev/null 2>&1 "Universal Ctags"; then \
+			ctags -R --languages=C --langmap=C:.h.c --c-kinds=+px \
+			    --fields=+S --fields-c=+"{properties}" --fields-c=+"{macrodef}" \
+			    --extras=+qr; \
+		else \
+			ctags -R --languages=C --langmap=C:.h.c --c-kinds=+px \
+			    --fields=+S; \
+		fi; \
+		printf "Created the Vim's style \`tags\` file.\\n"; \
+	else \
+	    printf 'Error: Please install a Ctags (e.g.: either the Exuberant or Universal %b' \
+	    'version) utility first.\n'; \
+	fi
 
 update-version:
 	@if test -x $(top_srcdir)/maint/utils/version.sh; then \
-- 
2.28.0

