diff --git a/misc/mc.menu b/misc/mc.menu
index 0cdcee4..e205719 100644
--- a/misc/mc.menu
+++ b/misc/mc.menu
@@ -284,10 +284,10 @@ B       Bzip2 or bunzip2 tagged files
         for i in %t
         do
           unset DECOMP
-	  case $i in 
+	  case "$i" in
 	    *.bz2) DECOMP=-d;;
 	  esac
-          bzip2 $DECOMP -v $i
+          bzip2 $DECOMP -v "$i"
         done
 
 + f \.tar.gz$ | f \.tgz$ | f \.tpz$ | f \.tar.Z$ | f \.tar.z$ | f \.tar.bz2$ | f \.tar.F$ & t r & ! t t
@@ -303,7 +303,7 @@ z       Extract compressed tar file to subdirectory
 	  *.tar.bz2) D="`basename %f .tar.bz2`"; set bunzip2 -c ;;
 	  *.tar.F) D="`basename %f .tar.F`"; set freeze -dc;
 	esac
-	mkdir $D; cd $D && ($1 $2 ../%f | tar xvf -)
+	mkdir "$D"; cd "$D" && ($1 $2 ../%f | tar xvf -)
 
 + t t
 Z       Extract compressed tar files to subdirectories
@@ -311,7 +311,7 @@ Z       Extract compressed tar files to subdirectories
         do
 	  set gzip -dc
           unset D
-	  case $i in
+	  case "$i" in
 	    *.tar.gz)  D="`basename $i .tar.gz`";;
 	    *.tgz)     D="`basename $i .tgz`";;
 	    *.tpz)     D="`basename $i .tpz`";;
@@ -320,7 +320,7 @@ Z       Extract compressed tar files to subdirectories
 	    *.tar.F)   D="`basename $i .tar.F`"; set freeze -dc;;
 	    *.tar.bz2) D="`basename $i .tar.bz2`"; set bunzip2 -c;;
           esac
-	  mkdir $D; (cd $D && $1 $2 ../$i | tar xvf -)
+	  mkdir "$D"; (cd "$D" && $1 $2 ../$i | tar xvf -)
         done
 
 + f \.gz$ | f \.tgz$ | f \.tpz$ | f \.Z$ | f \.z$ | f \.bz2$ & t r & ! t t
@@ -341,9 +341,9 @@ c       Convert gz<->bz2, tar.gz<->tar.bz2 & tgz->tar.bz2
 	  bz2)		D="`basename %f .bz2`";;
         esac
         if [ "$EXT" = "bz2" ]; then
-	    bunzip2 -v %f ; gzip -f9 -v $D
+	    bunzip2 -v %f ; gzip -f9 -v "$D"
 	else
-	    gunzip -v %f ; bzip2 -v $D
+	    gunzip -v %f ; bzip2 -v "$D"
 	fi
 
 + t t
@@ -353,7 +353,7 @@ C       Convert gz<->bz2, tar.gz<->tar.bz2 & tgz->tar.bz2
         do
 	  unset D
 	  unset EXT
-	  case $1 in
+	  case "$1" in
 	    *.tgz)	EXT=tgz;;
 	    *.tpz)	EXT=tpz;;
 	    *.Z)	EXT=Z;;
@@ -368,11 +368,11 @@ C       Convert gz<->bz2, tar.gz<->tar.bz2 & tgz->tar.bz2
 	      bz2)    D="`basename $1 .bz2`";;
 	  esac
 	  if [ "$EXT" = "bz2" ]; then
-	    bunzip2 -v $1
-	    gzip -f9 -v $D
+	    bunzip2 -v "$1"
+	    gzip -f9 -v "$D"
 	  else
-	    gunzip -v $1
-	    bzip2 -v $D
+	    gunzip -v "$1"
+	    bzip2 -v "$D"
 	  fi
 	  shift
         done

