diff --git a/configure.ac b/configure.ac
index 4b39f16..344e779 100644
--- a/configure.ac
+++ b/configure.ac
@@ -481,10 +481,15 @@ AC_SUBST(LIBS)
 dnl Libraries used only when building the mc binary
 AC_SUBST(MCLIBS)
 
-
 EXTHELPERSDIR=${prefix}/libexec/${PACKAGE}/ext.d
 AC_SUBST(EXTHELPERSDIR)
 
+EXTFSHELPERSDIR=${prefix}/libexec/${PACKAGE}/extfs.d
+AC_SUBST(EXTFSHELPERSDIR)
+
+FISHHELPERSDIR=${prefix}/libexec/${PACKAGE}/fish
+AC_SUBST(FISHHELPERSDIR)
+
 MAN_DATE="$(LC_ALL=C date "+%B %Y")"
 AC_SUBST(MAN_DATE)
 
diff --git a/src/vfs/extfs/helpers/Makefile.am b/src/vfs/extfs/helpers/Makefile.am
index 6e30e17..44ecb71 100644
--- a/src/vfs/extfs/helpers/Makefile.am
+++ b/src/vfs/extfs/helpers/Makefile.am
@@ -1,4 +1,4 @@
-extfsdir = $(libexecdir)/@PACKAGE@/extfs.d
+extfsdir = $(EXTFSHELPERSDIR)
 
 EXTFSCONFFILES = sfs.ini
 
diff --git a/src/vfs/fish/helpers/Makefile.am b/src/vfs/fish/helpers/Makefile.am
index 3064b01..5680f35 100644
--- a/src/vfs/fish/helpers/Makefile.am
+++ b/src/vfs/fish/helpers/Makefile.am
@@ -1,4 +1,4 @@
-fishdir = $(libexecdir)/@PACKAGE@/fish
+fishdir = $(FISHHELPERSDIR)
 
 # Files to install and distribute other than fish scripts
 FISH_MISC  = README.fish
