diff -Nurwd orig/mc-4.6.2~git20080311/vfs/extfs/u7z mc-4.6.2~git20080311/vfs/extfs/u7z
--- orig/mc-4.6.2~git20080311/vfs/extfs/u7z	2008-07-26 11:26:48.000000000 +0200
+++ mc-4.6.2~git20080311/vfs/extfs/u7z	2008-07-26 11:30:24.000000000 +0200
@@ -45,8 +45,8 @@
 	#first we check if we have old p7zip archive with prefix ./ in filename
 	$P7ZIP l "$1" "$2" | grep -q "0 files" && \
 	EXFNAME=*./"$2" || EXFNAME="$2"
-	$P7ZIP d "$1" "$EXFNAME" 2>&1 | grep -q E_NOTIMPL &> /dev/null && \
-	{ echo -e "Function not implemented...\n7z cannot delete from solid archive." >&2 ; exit 1 ; }
+	$P7ZIP d "$1" "$EXFNAME" 2>&1 | grep -q E_NOTIMPL > /dev/null 2>&1 && \
+	{ printf "Function not implemented...\n7z cannot delete from solid archive.\n" >&2 ; exit 1 ; }
 }
 
 mcu7zip_rmdir ()
@@ -54,8 +54,8 @@
 	#first we check if we have old p7zip archive with prefix ./ in filename
 	$P7ZIP l "$1" "$2" | grep -q "0 files" && \
 	EXFNAME=*./"$2" || EXFNAME="$2"
-	$P7ZIP d "$1" "$EXFNAME"/ 2>&1 | grep -q E_NOTIMPL &> /dev/null && \
-	{ echo -e "Function not implemented...\n7z cannot delete from solid archive." >&2 ; exit 1 ; }
+	$P7ZIP d "$1" "$EXFNAME"/ 2>&1 | grep -q E_NOTIMPL > /dev/null 2>&1  && \
+	{ printf "Function not implemented...\n7z cannot delete from solid archive.\n" >&2 ; exit 1 ; }
 }
 
 # override any locale for dates
diff -Nurwd orig/mc-4.6.2~git20080311/vfs/extfs/uarj.in mc-4.6.2~git20080311/vfs/extfs/uarj.in
--- orig/mc-4.6.2~git20080311/vfs/extfs/uarj.in	2008-07-26 11:26:48.000000000 +0200
+++ mc-4.6.2~git20080311/vfs/extfs/uarj.in	2008-07-26 11:34:19.000000000 +0200
@@ -16,7 +16,7 @@
 
 mcarjfs_list ()
 {
- $ARJ v "$1" | @AWK@ -v uuid=${UID-0} '
+ $ARJ v "$1" | @AWK@ -v uuid=$(id -ru) '
   { 
    if (($0 ~ /^[0-9]+\) .*/)||($0 ~ /^------------ ---------- ---------- -----/)){
     if (filestr ~ /^[0-9]+\) .*/) {
diff -Nurwd orig/mc-4.6.2~git20080311/vfs/extfs/uha.in mc-4.6.2~git20080311/vfs/extfs/uha.in
--- orig/mc-4.6.2~git20080311/vfs/extfs/uha.in	2008-07-26 11:26:48.000000000 +0200
+++ mc-4.6.2~git20080311/vfs/extfs/uha.in	2008-07-26 11:34:39.000000000 +0200
@@ -10,7 +10,7 @@
 
 mchafs_list ()
 {
-    $HA lf "$1" 2>/dev/null | @AWK@ -v uid=${UID-0} '
+    $HA lf "$1" 2>/dev/null | @AWK@ -v uid=$(id -ru)  '
 /^===========/ {next}
 {
  if ($5="%" && $8~/DIR|ASC|HSC|CPY/) {
diff -Nurwd orig/mc-4.6.2~git20080311/vfs/extfs/uzoo.in mc-4.6.2~git20080311/vfs/extfs/uzoo.in
--- orig/mc-4.6.2~git20080311/vfs/extfs/uzoo.in	2008-07-26 11:26:48.000000000 +0200
+++ mc-4.6.2~git20080311/vfs/extfs/uzoo.in	2008-07-26 11:36:00.000000000 +0200
@@ -23,7 +23,7 @@
 mczoofs_list ()
 {
     mklink "$1"
-    $ZOO lq "$ARCHIVE" | @AWK@ -v uid=${UID-0} '
+    $ZOO lq "$ARCHIVE" | @AWK@ -v uid=$(id -ru)  '
 /^[^\ ]/ { next }
 {
 if (NF < 8)
