diff -ur mc-4.8.1.orig/m4.include/mc-get-fs-info.m4 mc-4.8.1/m4.include/mc-get-fs-info.m4
--- mc-4.8.1.orig/m4.include/mc-get-fs-info.m4	2011-12-12 10:15:38.000000000 +0100
+++ mc-4.8.1/m4.include/mc-get-fs-info.m4	2011-12-16 12:42:10.000000000 +0100
@@ -5,7 +5,7 @@
 
 AC_DEFUN([gl_POSIX_FALLOCATE], [
     dnl * Old glibcs have broken posix_fallocate(). Make sure not to use it.
-    AC_TRY_COMPILE([
+    AC_TRY_LINK([
         #define _XOPEN_SOURCE 600
         #include <stdlib.h>
         #if defined(__GLIBC__) && (__GLIBC__ < 2 || __GLIBC_MINOR__ < 7)
