--- misc/mc.ext.in.orig	2014-03-13 10:06:03.000000000 +0100
+++ misc/mc.ext.in	2014-10-17 10:33:35.843428924 +0200
@@ -655,6 +655,9 @@
 # Gettext Catalogs
 shell/.mo
 	View=%view{ascii} @EXTHELPERSDIR@/misc.sh view mo
+shell/.po
+	View=%view{ascii} @EXTHELPERSDIR@/misc.sh view po
+	Open=@EXTHELPERSDIR@/misc.sh open po
 
 # lyx
 shell/i/.lyx
--- misc/ext.d/misc.sh.in.orig	2014-03-28 10:22:39.000000000 +0100
+++ misc/ext.d/misc.sh.in	2014-10-17 10:24:16.680236084 +0200
@@ -41,6 +41,9 @@
         msgunfmt "${MC_EXT_FILENAME}" || \
             cat "${MC_EXT_FILENAME}"
         ;;
+    po)
+        poedit "${MC_EXT_FILENAME}"
+        ;;
     lyx)
         lyxcat "${MC_EXT_FILENAME}"
         ;;
@@ -75,6 +78,9 @@
             (glade-2 "${MC_EXT_FILENAME}" >/dev/null 2>&1 &)
         fi
         ;;
+    po)
+        poedit "${MC_EXT_FILENAME}"
+        ;;
     lyx)
         lyx "${MC_EXT_FILENAME}"
         ;;
