diff -urN mc-4.7.5.5.orig/src/viewer/datasource.c mc-4.7.5.5/src/viewer/datasource.c
--- mc-4.7.5.5.orig/src/viewer/datasource.c	2011-09-28 12:48:36.000000000 -0400
+++ mc-4.7.5.5/src/viewer/datasource.c	2011-10-14 06:04:01.000000000 -0400
@@ -164,6 +164,7 @@
     int res = -1;
     gunichar ch;
     gchar *next_ch = NULL;
+    gchar utf8buf[6 + 1];
 
     *char_width = 0;
     *result = FALSE;
@@ -189,6 +190,20 @@
 
     res = g_utf8_get_char_validated (str, -1);
 
+    if (res < 0)
+    {
+        /* Retry with explicit bytes to make sure it's not a buffer boundary */
+        int i;
+        for (i = 0; i < 6; i++)
+            if (mcview_get_byte (view, byte_index + i, &res))
+                utf8buf[i] = res;
+            else
+                utf8buf[i] = '\0';
+        utf8buf[6] = '\0';
+        str = utf8buf;
+        res = g_utf8_get_char_validated (str, -1);
+    }
+
     if (res < 0)
     {
         ch = *str;
