--- src/subshell/common.c
+++ src/subshell/common.c
@@ -908,7 +908,7 @@ init_subshell_precmd (char *precmd, size_t buff_size)
  * replaced by the backslash-escape sequence \0nnn, where "nnn" is the
  * numeric value of the character converted to octal number.
  * 
- *   cd "`printf "%b" 'ABC\0nnnDEF\0nnnXYZ'`"
+ *   cd "`printf %b 'ABC\0nnnDEF\0nnnXYZ'`"
  *
  */
 
@@ -921,7 +921,7 @@ subshell_name_quote (const char *s)
 
     if (mc_global.shell->type == SHELL_FISH)
     {
-        quote_cmd_start = "(printf \"%b\" '";
+        quote_cmd_start = "(printf %b '";
         quote_cmd_end = "')";
     }
     /* TODO: When BusyBox printf is fixed, get rid of this "else if", see
@@ -933,7 +933,7 @@ subshell_name_quote (const char *s)
        } */
     else
     {
-        quote_cmd_start = "\"`printf \"%b\" '";
+        quote_cmd_start = "\"`printf %b '";
         quote_cmd_end = "'`\"";
     }
 
