--- lib/widget/listbox.c-orig	2016-06-07 18:17:03.000000000 +0200
+++ lib/widget/listbox.c	2016-06-07 21:05:02.001011162 +0200
@@ -52,7 +52,7 @@
 /*** file scope macro definitions ****************************************************************/
 
 /* Gives the position of the last item. */
-#define LISTBOX_LAST(l) ((l) && ((l)->list) && g_queue_is_empty ((l)->list) ? 0 : (int) g_queue_get_length ((l)->list) - 1)
+#define LISTBOX_LAST(l) (l == NULL || l->list == NULL || g_queue_is_empty ((l)->list) ? 0 : (int) g_queue_get_length ((l)->list) - 1)
 
 /*** file scope type declarations ****************************************************************/
 
@@ -229,10 +229,13 @@
 static void
 listbox_fwd (WListbox * l, gboolean wrap)
 {
+	if (l->list != NULL)
+	{
     if ((guint) l->pos + 1 < g_queue_get_length (l->list))
         listbox_select_entry (l, l->pos + 1);
     else if (wrap)
         listbox_select_first (l);
+	}
 }
 
 /* --------------------------------------------------------------------------------------------- */
@@ -400,6 +403,9 @@
 {
     int action;
 
+	if (l->list == NULL)
+		return;
+
     if (l->callback != NULL)
         action = l->callback (l);
     else
