--- uzip.orig	2010-03-20 22:11:55.000000000 +0300
+++ uzip	2010-03-24 19:18:46.000000000 +0300
@@ -262,7 +262,8 @@
 			my @match = /$regex_nonzipinfo_line/;
 			next if ($#match != 10);
 			my @rmatch = ('', '', 'unknown', $match[0], '', $match[2], $match[1],
-					$match[6] + ($match[6] < 70 ? 2000 : 1900), $match[4], $match[5],
+					$match[6] > 100 ? $match[6] : $match[6] + ($match[6] < 70 ? 2000 : 1900)
+					$match[4], $match[5],
 					$match[7], $match[8], "00", $match[10]);
 			&checked_print_file(@rmatch);
 		}
