diff -ur mc-4.8.1.3.orig/misc/skins/sand256.ini mc-4.8.1.3/misc/skins/sand256.ini
--- mc-4.8.1.3.orig/misc/skins/sand256.ini	2012-04-25 18:10:07.000000000 +0200
+++ mc-4.8.1.3/misc/skins/sand256.ini	2012-05-11 10:28:01.000000000 +0200
@@ -131,6 +131,11 @@
     menuhotsel=;rgb551;underline
     menuinactive=
 
+[popupmenu]
+    _default_=black;rgb553 
+    menusel=;rgb452;underline
+    menutitle=;;underline
+
 [buttonbar]
     hotkey=black;rgb554
     button=black;rgb452
