From bd457ec7804b82b7383cb60ba88580f9a51ea261 Mon Sep 17 00:00:00 2001
From: Vit Rosin <vit_r@list.ru>
Date: Sat, 28 Nov 2009 12:40:48 +0000
Subject: [PATCH] UNneeded_goto

---
 edit/edit.c |   43 ++++++++++++++++++-------------------------
 1 files changed, 18 insertions(+), 25 deletions(-)

diff --git a/edit/edit.c b/edit/edit.c
index 0b0b1eb..04528b8 100644
--- a/edit/edit.c
+++ b/edit/edit.c
@@ -3276,47 +3276,40 @@ edit_execute_macro (WEdit *edit, struct macro macro[], int n)
     edit->macro_depth--;
 }
 
-/* User edit menu, like user menu (F2) but only in editor. */
+/* On file selected (i.e. by arrows) (F4) invokes the editor
+   then (F11) or (S-F1) invokes user menu */
 static void
 user_menu (WEdit * edit)
 {
     FILE *fd;
-    int nomark;
-    struct stat status;
+    int nomark, rc = 0;
     long start_mark, end_mark;
+    struct stat status;
     char *block_file = concat_dir_and_file (home_dir, EDIT_BLOCK_FILE);
-    int rc = 0;
 
+    /* remember marked or not */
     nomark = eval_marks (edit, &start_mark, &end_mark);
-    if (!nomark)		/* remember marked or not */
-	edit_save_block (edit, block_file, start_mark, end_mark);
+    if (!nomark)
+        edit_save_block (edit, block_file, start_mark, end_mark);
 
     /* run shell scripts from menu */
     user_menu_cmd (edit);
 
-    if (mc_stat (block_file, &status) != 0 || !status.st_size) {
-	/* no block messages */
-	goto cleanup;
-    }
+    if (mc_stat (block_file, &status) == 0 || status.st_size) {
+        /* there are block messages */
+        if (!nomark) /* i.e. we have marked block */
+            rc = edit_block_delete_cmd (edit);
 
-    if (!nomark) {
-	/* i.e. we have marked block */
-	rc = edit_block_delete_cmd (edit);
-    }
+        if (!rc) 
+            edit_insert_file (edit, block_file);
 
-    if (!rc) {
-	edit_insert_file (edit, block_file);
-    }
+        fd = fopen (block_file, "w");
+        if (fd) /* truncate block file */
+            fclose (fd);
 
-    /* truncate block file */
-    if ((fd = fopen (block_file, "w"))) {
-	fclose (fd);
+        edit_refresh_cmd (edit);
+        edit->force |= REDRAW_COMPLETELY;
     }
-
-    edit_refresh_cmd (edit);
-    edit->force |= REDRAW_COMPLETELY;
-
-cleanup:
     g_free (block_file);
 }
 
-- 
1.6.4

