diff --git a/lib/tty/tty-ncurses.c b/lib/tty/tty-ncurses.c
index 7481a3e..6a66342 100644
--- a/lib/tty/tty-ncurses.c
+++ b/lib/tty/tty-ncurses.c
@@ -81,25 +81,6 @@ static int mc_curs_row, mc_curs_col;
 /*** file scope functions ************************************************************************/
 /* --------------------------------------------------------------------------------------------- */
 
-/* --------------------------------------------------------------------------------------------- */
-
-static void
-tty_setup_sigwinch (void (*handler) (int))
-{
-#if (NCURSES_VERSION_MAJOR >= 4) && defined (SIGWINCH)
-    struct sigaction act, oact;
-    act.sa_handler = handler;
-    sigemptyset (&act.sa_mask);
-    act.sa_flags = 0;
-#ifdef SA_RESTART
-    act.sa_flags |= SA_RESTART;
-#endif /* SA_RESTART */
-    sigaction (SIGWINCH, &act, &oact);
-#endif /* SIGWINCH */
-}
-
-/* --------------------------------------------------------------------------------------------- */
-
 static void
 sigwinch_handler (int dummy)
 {
@@ -205,8 +186,6 @@ tty_init (gboolean mouse_enable, gboolean is_xterm)
     noecho ();
     keypad (stdscr, TRUE);
     nodelay (stdscr, FALSE);
-
-    tty_setup_sigwinch (sigwinch_handler);
 }
 
 /* --------------------------------------------------------------------------------------------- */
@@ -630,3 +609,21 @@ tty_beep (void)
 }
 
 /* --------------------------------------------------------------------------------------------- */
+
+void
+tty_setup_sigwinch (void)
+{
+#if (NCURSES_VERSION_MAJOR >= 4) && defined (SIGWINCH)
+    struct sigaction act, oact;
+
+    act.sa_handler = sigwinch_handler;
+    sigemptyset (&act.sa_mask);
+    act.sa_flags = 0;
+#ifdef SA_RESTART
+    act.sa_flags |= SA_RESTART;
+#endif /* SA_RESTART */
+    sigaction (SIGWINCH, &act, &oact);
+#endif /* SIGWINCH */
+}
+
+/* --------------------------------------------------------------------------------------------- */
diff --git a/lib/tty/tty-slang.c b/lib/tty/tty-slang.c
index 6b7fd32..cae30c7 100644
--- a/lib/tty/tty-slang.c
+++ b/lib/tty/tty-slang.c
@@ -132,23 +132,6 @@ static const struct
 /* --------------------------------------------------------------------------------------------- */
 
 static void
-tty_setup_sigwinch (void (*handler) (int))
-{
-#ifdef SIGWINCH
-    struct sigaction act, oact;
-    act.sa_handler = handler;
-    sigemptyset (&act.sa_mask);
-    act.sa_flags = 0;
-#ifdef SA_RESTART
-    act.sa_flags |= SA_RESTART;
-#endif /* SA_RESTART */
-    sigaction (SIGWINCH, &act, &oact);
-#endif /* SIGWINCH */
-}
-
-/* --------------------------------------------------------------------------------------------- */
-
-static void
 sigwinch_handler (int dummy)
 {
     (void) dummy;
@@ -342,8 +325,6 @@ tty_init (gboolean mouse_enable, gboolean is_xterm)
     do_enter_ca_mode ();
     tty_keypad (TRUE);
     tty_nodelay (FALSE);
-
-    tty_setup_sigwinch (sigwinch_handler);
 }
 
 /* --------------------------------------------------------------------------------------------- */
@@ -737,3 +718,21 @@ tty_beep (void)
 }
 
 /* --------------------------------------------------------------------------------------------- */
+
+void
+tty_setup_sigwinch (void)
+{
+#ifdef SIGWINCH
+    struct sigaction act, oact;
+
+    act.sa_handler = sigwinch_handler;
+    sigemptyset (&act.sa_mask);
+    act.sa_flags = 0;
+#ifdef SA_RESTART
+    act.sa_flags |= SA_RESTART;
+#endif /* SA_RESTART */
+    sigaction (SIGWINCH, &act, &oact);
+#endif /* SIGWINCH */
+}
+
+/* --------------------------------------------------------------------------------------------- */
diff --git a/lib/tty/tty.h b/lib/tty/tty.h
index 0450208..5df0ff4 100644
--- a/lib/tty/tty.h
+++ b/lib/tty/tty.h
@@ -130,6 +130,7 @@ extern void tty_fill_region (int y, int x, int rows, int cols, unsigned char ch)
 extern int tty_resize (int fd);
 extern void tty_refresh (void);
 extern void tty_change_screen_size (void);
+extern void tty_setup_sigwinch (void);
 
 extern int mc_tty_normalize_lines_char (const char *);
 
diff --git a/src/filemanager/midnight.c b/src/filemanager/midnight.c
index 0ba2fd8..0ed4f04 100644
--- a/src/filemanager/midnight.c
+++ b/src/filemanager/midnight.c
@@ -877,6 +877,8 @@ setup_mc (void)
         add_select_channel (mc_global.tty.subshell_pty, load_prompt, 0);
 #endif /* !HAVE_SUBSHELL_SUPPORT */
 
+    tty_setup_sigwinch ();
+
     if ((tty_baudrate () < 9600) || mc_global.tty.slow_terminal)
         verbose = 0;
 }
