--- editdraw.c.old	2005-09-05 17:35:26.000000000 +0200
+++ editdraw.c	2006-01-30 16:25:56.000000000 +0100
@@ -304,7 +304,8 @@
 	if (row <= edit->total_lines - edit->start_line) {
 	    if (use_colors) {
 		tws = edit_eol (edit, b);
-		while (tws > b && edit_get_byte (edit, tws - 1) == ' ')
+		while (tws > b && ((c = edit_get_byte (edit, tws - 1)) == ' '
+				   || c == '\t'))
 		    tws--;
 	    }
 
