From 4826af732849a19c4cdfea999a511a8a7b4156c4 Mon Sep 17 00:00:00 2001
From: Mooffie <mooffie@gmail.com>
Date: Thu, 26 Nov 2015 00:30:56 +0200
Subject: [PATCH] Fix WButton to use the new mouse API.

---
 lib/widget/button.c | 31 ++++++++++++++++---------------
 1 file changed, 16 insertions(+), 15 deletions(-)

diff --git a/lib/widget/button.c b/lib/widget/button.c
index cf7de6d..5f4eb5e 100644
--- a/lib/widget/button.c
+++ b/lib/widget/button.c
@@ -174,25 +174,25 @@ button_callback (Widget * w, Widget * sender, widget_msg_t msg, int parm, void *
 
 /* --------------------------------------------------------------------------------------------- */
 
-static int
-button_event (Gpm_Event * event, void *data)
+static void
+button_mouse_callback (Widget * w, mouse_msg_t msg, mouse_event_t * event)
 {
-    Widget *w = WIDGET (data);
+    (void) event;
 
-    if (!mouse_global_in_widget (event, w))
-        return MOU_UNHANDLED;
-
-    if ((event->type & (GPM_DOWN | GPM_UP)) != 0)
+    switch (msg)
     {
+    case MSG_MOUSE_DOWN:
         dlg_select_widget (w);
-        if ((event->type & GPM_UP) != 0)
-        {
-            send_message (w, NULL, MSG_KEY, ' ', NULL);
-            send_message (w->owner, w, MSG_POST_KEY, ' ', NULL);
-        }
-    }
+        break;
 
-    return MOU_NORMAL;
+    case MSG_MOUSE_CLICK:
+        send_message (w, NULL, MSG_KEY, ' ', NULL);
+        send_message (w->owner, w, MSG_POST_KEY, ' ', NULL);
+        break;
+
+    default:
+        break;
+    }
 }
 
 /* --------------------------------------------------------------------------------------------- */
@@ -211,7 +211,8 @@ button_new (int y, int x, int action, button_flags_t flags, const char *text, bc
     b->action = action;
     b->flags = flags;
     b->text = parse_hotkey (text);
-    widget_init (w, y, x, 1, button_get_len (b), button_callback, button_event);
+    widget_init (w, y, x, 1, button_get_len (b), button_callback, NULL);
+    set_easy_mouse_callback (w, button_mouse_callback);
     b->selected = FALSE;
     b->callback = callback;
     widget_want_hotkey (w, TRUE);
-- 
2.1.4

