diff -Naur mc-4.7.0-pre1~/edit/edit-widget.h mc-4.7.0-pre1/edit/edit-widget.h
--- mc-4.7.0-pre1~/edit/edit-widget.h	2009-07-20 21:03:45.000000000 +0300
+++ mc-4.7.0-pre1/edit/edit-widget.h	2009-08-07 17:00:13.000000000 +0300
@@ -62,6 +62,7 @@
     int replace_mode;
     int replace_backwards;
     int replace_case;
+    int whole_words;
     int all_codepages;
 
     long search_start;		/* First character to start searching from */
diff -Naur mc-4.7.0-pre1~/edit/editcmd.c mc-4.7.0-pre1/edit/editcmd.c
--- mc-4.7.0-pre1~/edit/editcmd.c	2009-07-31 19:36:25.000000000 +0300
+++ mc-4.7.0-pre1/edit/editcmd.c	2009-08-07 17:01:01.000000000 +0300
@@ -1407,6 +1407,7 @@
 	edit->search->search_type = edit->search_type;
 	edit->search->is_all_charsets = edit->all_codepages;
 	edit->search->is_case_sentitive = edit->replace_case;
+	edit->search->whole_words = edit->whole_words;
 	edit->search->search_fn = edit_search_cmd_callback;
     }
 
@@ -1604,6 +1605,7 @@
 	edit->search->search_type = edit->search_type;
 	edit->search->is_all_charsets = edit->all_codepages;
 	edit->search->is_case_sentitive = edit->replace_case;
+	edit->search->whole_words = edit->whole_words;
 	edit->search->search_fn = edit_search_cmd_callback;
     }
 
diff -Naur mc-4.7.0-pre1~/edit/editcmd_dialogs.c mc-4.7.0-pre1/edit/editcmd_dialogs.c
--- mc-4.7.0-pre1~/edit/editcmd_dialogs.c	2009-07-31 19:36:25.000000000 +0300
+++ mc-4.7.0-pre1/edit/editcmd_dialogs.c	2009-08-07 17:01:33.000000000 +0300
@@ -49,7 +49,7 @@
 /*** file scope macro definitions **************************************/
 
 #define SEARCH_DLG_WIDTH 58
-#define SEARCH_DLG_MIN_HEIGHT 11
+#define SEARCH_DLG_MIN_HEIGHT 12
 #define SEARCH_DLG_HEIGHT_SUPPLY 3
 
 #define REPLACE_DLG_WIDTH 58
@@ -84,6 +84,7 @@
 {
     int treplace_backwards = edit->replace_backwards;
     int treplace_case = edit->replace_case;
+    int twhole_words = edit->whole_words;
     int tall_codepages = edit->all_codepages;
     mc_search_type_t ttype_of_search = edit->search_type;
     int dialog_result;
@@ -101,13 +102,16 @@
          0, 0, NULL, NULL, NULL},
 
 #ifdef HAVE_CHARSET
-        {quick_checkbox, 33, REPLACE_DLG_WIDTH, 9, REPLACE_DLG_HEIGHT, N_("All charsets"), 0, 0,
+        {quick_checkbox, 33, REPLACE_DLG_WIDTH, 10, REPLACE_DLG_HEIGHT, N_("All charsets"), 0, 0,
          &tall_codepages, 0, NULL, NULL, NULL},
 #endif
 
-        {quick_checkbox, 33, REPLACE_DLG_WIDTH, 8, REPLACE_DLG_HEIGHT, N_("&Backwards"), 0, 0,
+        {quick_checkbox, 33, REPLACE_DLG_WIDTH, 9, REPLACE_DLG_HEIGHT, N_("&Backwards"), 0, 0,
          &treplace_backwards, 0, NULL, NULL, NULL},
 
+        {quick_checkbox, 33, REPLACE_DLG_WIDTH, 8, REPLACE_DLG_HEIGHT, N_("&Whole words"), 0, 0,
+         &twhole_words, 0, NULL, NULL, NULL},
+
         {quick_checkbox, 33, REPLACE_DLG_WIDTH, 7, REPLACE_DLG_HEIGHT, N_("case &Sensitive"), 0, 0,
          &treplace_case, 0, NULL, NULL, NULL},
 
@@ -153,6 +157,7 @@
         edit->all_codepages = tall_codepages;
         edit->replace_backwards = treplace_backwards;
         edit->replace_case = treplace_case;
+        edit->whole_words = twhole_words;
         return;
     } else {
         *replace_text = NULL;
@@ -167,6 +172,7 @@
 editcmd_dialog_search_show (WEdit * edit, char **search_text)
 {
     int tsearch_case = edit->replace_case;
+    int twhole_words = edit->whole_words;
     int tsearch_backwards = edit->replace_backwards;
     int tall_codepages = edit->all_codepages;
     mc_search_type_t ttype_of_search = edit->search_type;
@@ -175,18 +181,20 @@
 
 
     QuickWidget quick_widgets[] = {
-        {quick_button, 6, 10, 9, SEARCH_DLG_HEIGHT, N_("&Cancel"), 0, B_CANCEL, 0,
+        {quick_button, 6, 10, 10, SEARCH_DLG_HEIGHT, N_("&Cancel"), 0, B_CANCEL, 0,
          0, NULL, NULL, NULL},
-        {quick_button, 2, 10, 9, SEARCH_DLG_HEIGHT, N_("&OK"), 0, B_ENTER, 0,
+        {quick_button, 2, 10, 10, SEARCH_DLG_HEIGHT, N_("&OK"), 0, B_ENTER, 0,
          0, NULL, NULL, NULL},
 
 #ifdef HAVE_CHARSET
-        {quick_checkbox, 33, SEARCH_DLG_WIDTH, 7, SEARCH_DLG_HEIGHT, N_("All charsets"), 0, 0,
+        {quick_checkbox, 33, SEARCH_DLG_WIDTH, 8, SEARCH_DLG_HEIGHT, N_("All charsets"), 0, 0,
          &tall_codepages, 0, NULL, NULL, NULL},
 #endif
 
-        {quick_checkbox, 33, SEARCH_DLG_WIDTH, 6, SEARCH_DLG_HEIGHT, N_("&Backwards"), 0, 0,
+        {quick_checkbox, 33, SEARCH_DLG_WIDTH, 7, SEARCH_DLG_HEIGHT, N_("&Backwards"), 0, 0,
          &tsearch_backwards, 0, NULL, NULL, NULL},
+        {quick_checkbox, 33, SEARCH_DLG_WIDTH, 6, SEARCH_DLG_HEIGHT, N_("&Whole words"), 0, 0,
+         &twhole_words, 0, NULL, NULL, NULL},
         {quick_checkbox, 33, SEARCH_DLG_WIDTH, 5, SEARCH_DLG_HEIGHT, N_("case &Sensitive"), 0, 0,
          &tsearch_case, 0, NULL, NULL, NULL},
 
@@ -213,6 +221,7 @@
         edit->replace_backwards = tsearch_backwards;
         edit->all_codepages = tall_codepages;
         edit->replace_case = tsearch_case;
+        edit->whole_words = twhole_words;
     } else {
         *search_text = NULL;
     }
diff -Naur mc-4.7.0-pre1~/src/find.c mc-4.7.0-pre1/src/find.c
--- mc-4.7.0-pre1~/src/find.c	2009-07-28 09:43:41.000000000 +0300
+++ mc-4.7.0-pre1/src/find.c	2009-08-07 17:02:47.000000000 +0300
@@ -89,6 +89,7 @@
 static WCheck *file_case_sens;		/* "case sensitive" checkbox */
 static WCheck *find_regex_cbox;		/* [x] find regular expression */
 static WCheck *content_case_sens;	/* "case sensitive" checkbox */
+static WCheck *content_whole_words_cbox;/* "whole words" checkbox */
 
 static int running = 0;			/* nice flag */
 static char *find_pattern = NULL;	/* Pattern to search */
@@ -148,6 +149,7 @@
 static gboolean content_regexp_flag = FALSE;
 static gboolean content_all_charsets_flag = FALSE;
 static gboolean content_case_sensitive = TRUE;
+static gboolean content_whole_words = FALSE;
 
 static int find_recursively = 1;
 
@@ -261,6 +263,7 @@
 
     static const char *file_case_label = N_("Cas&e sensitive");
     static const char *content_case_label = N_("Case sens&itive");
+    static const char *content_whole_words_label = N_("&Whole words");
     static const char *file_recurs_label = N_("&Find recursively");
     static const char *file_skip_hidden_label = N_("S&kip hidden");
     static const char *file_all_charsets_label = N_("&All charsets");
@@ -327,14 +330,17 @@
 		button_new (FIND_Y - 3, FIND_X/4 - b0/2, B_ENTER, DEFPUSH_BUTTON, buts[0], 0));
 
 #ifdef HAVE_CHARSET
-    content_all_charsets_cbox = check_new (9, FIND_X / 2 + 1,
+    content_all_charsets_cbox = check_new (11, FIND_X / 2 + 1,
 		content_all_charsets_flag, content_all_charsets_label);
     add_widget (find_dlg, content_all_charsets_cbox);
 #endif
 
-    find_regex_cbox = check_new (8, FIND_X / 2 + 1, content_regexp_flag, content_regexp_label);
+    find_regex_cbox = check_new (9, FIND_X / 2 + 1, content_regexp_flag, content_regexp_label);
     add_widget (find_dlg, find_regex_cbox);
 
+    content_whole_words_cbox = check_new (8, FIND_X / 2 + 1, content_whole_words, content_whole_words_label);
+    add_widget (find_dlg, content_whole_words_cbox);
+
     content_case_sens = check_new (7, FIND_X / 2 + 1, content_case_sensitive, content_case_label);
     add_widget (find_dlg, content_case_sens);
 
@@ -385,6 +391,7 @@
 	content_all_charsets_flag = content_all_charsets_cbox->state & C_BOOL;
 #endif
 	content_case_sensitive = content_case_sens->state & C_BOOL;
+	content_whole_words = content_whole_words_cbox->state & C_BOOL;
 	content_regexp_flag = find_regex_cbox->state & C_BOOL;
 	file_pattern_flag = file_pattern_cbox->state & C_BOOL;
 	file_case_sensitive = file_case_sens->state & C_BOOL;
@@ -418,6 +425,7 @@
 	content_all_charsets_flag = content_all_charsets_cbox->state & C_BOOL;
 #endif
 	content_case_sensitive = content_case_sens->state & C_BOOL;
+	content_whole_words = content_whole_words_cbox->state & C_BOOL;
 	content_regexp_flag = find_regex_cbox->state & C_BOOL;
 	find_recursively = recursively_cbox->state & C_BOOL;
 	file_pattern_flag = file_pattern_cbox->state & C_BOOL;
@@ -741,6 +749,7 @@
     {
         search_content_handle->search_type = (content_regexp_flag) ? MC_SEARCH_T_REGEX : MC_SEARCH_T_NORMAL;
         search_content_handle->is_case_sentitive = content_case_sensitive;
+        search_content_handle->whole_words = content_whole_words;
         search_content_handle->is_all_charsets = content_all_charsets_flag;
     }
     search_file_handle = mc_search_new(find_pattern, -1);
diff -Naur mc-4.7.0-pre1~/src/view.c mc-4.7.0-pre1/src/view.c
--- mc-4.7.0-pre1~/src/view.c	2009-07-31 19:36:26.000000000 +0300
+++ mc-4.7.0-pre1/src/view.c	2009-08-07 17:03:18.000000000 +0300
@@ -261,6 +261,7 @@
     mc_search_type_t search_type;
     gboolean search_all_codepages;
     gboolean search_case;
+    gboolean whole_words;
     gboolean search_backwards;
 
     int search_numNeedSkipChar;
@@ -3146,7 +3147,7 @@
     if (view->search_backwards) {
         search_end = view_get_filesize (view);
         while ((int) search_start >= 0) {
-            if (search_end - search_start > view->search->original_len && mc_search_is_fixed_search_str(view->search))
+            if (search_end > search_start + view->search->original_len && mc_search_is_fixed_search_str(view->search))
                 search_end = search_start + view->search->original_len;
 
             view_read_start (view, &view->search_onechar_info, search_start);
@@ -3268,6 +3269,7 @@
     int ttype_of_search = (int) view->search_type;
     int tall_codepages = (int) view->search_all_codepages;
     int tsearch_case = (int) view->search_case;
+    int twhole_words = (int) view->whole_words;
     int tsearch_backwards = (int) view->search_backwards;
 
     gchar **list_of_types = mc_search_get_types_strings_array();
@@ -3282,13 +3284,16 @@
 	 0, 0, NULL, NULL, NULL},
 
 #ifdef HAVE_CHARSET
-        {quick_checkbox, SEARCH_DLG_WIDTH/2 + 3, SEARCH_DLG_WIDTH, 6, SEARCH_DLG_HEIGHT, N_("All charsets"), 0, 0,
+        {quick_checkbox, SEARCH_DLG_WIDTH/2 + 3, SEARCH_DLG_WIDTH, 7, SEARCH_DLG_HEIGHT, N_("All charsets"), 0, 0,
          &tall_codepages, 0, NULL, NULL, NULL},
 #endif
 
-	{quick_checkbox, SEARCH_DLG_WIDTH/2 + 3, SEARCH_DLG_WIDTH, 5, SEARCH_DLG_HEIGHT,
+	{quick_checkbox, SEARCH_DLG_WIDTH/2 + 3, SEARCH_DLG_WIDTH, 6, SEARCH_DLG_HEIGHT,
 	 N_("&Backwards"), 0, 0, &tsearch_backwards, 0, NULL, NULL, NULL},
 
+        {quick_checkbox, SEARCH_DLG_WIDTH/2 + 3, SEARCH_DLG_WIDTH, 5, SEARCH_DLG_HEIGHT, N_("&Whole words"), 0, 0,
+         &twhole_words, 0, NULL, NULL, NULL},
+
         {quick_checkbox, SEARCH_DLG_WIDTH/2 + 3, SEARCH_DLG_WIDTH, 4, SEARCH_DLG_HEIGHT, N_("case &Sensitive"), 0, 0,
          &tsearch_case, 0, NULL, NULL, NULL},
 
@@ -3321,6 +3326,7 @@
 
     view->search_all_codepages = (gboolean) tall_codepages;
     view->search_case = (gboolean) tsearch_case;
+    view->whole_words = (gboolean) twhole_words;
 
     if (exp == NULL || exp[0] == '\0')
 	goto cleanup;
@@ -3349,6 +3355,7 @@
     view->search->search_type = view->search_type;
     view->search->is_all_charsets = view->search_all_codepages;
     view->search->is_case_sentitive = view->search_case;
+    view->search->whole_words = view->whole_words;
     view->search->search_fn = view_search_cmd_callback;
     view->search->update_fn = view_search_update_cmd_callback;
 
