diff --git a/src/viewer/actions_cmd.c b/src/viewer/actions_cmd.c
index c4d52b7..7e9655c 100644
--- a/src/viewer/actions_cmd.c
+++ b/src/viewer/actions_cmd.c
@@ -254,7 +254,7 @@ mcview_handle_editkey (mcview_t * view, int key)
     else
     {
         /* Text editing */
-        if (key < 256 && ((key == '\n') || is_printable (key)))
+        if (key < 256 && key != '\t')
             byte_val = key;
         else
             return MSG_NOT_HANDLED;
