From ee570d374618dabd9cd30018099254e7b1be9e3b Mon Sep 17 00:00:00 2001
From: Mooffie <mooffie@gmail.com>
Date: Sun, 24 Jan 2016 15:32:33 +0200
Subject: [PATCH] "Listing mode" dialog: remove old code.

---
 src/filemanager/boxes.c | 62 -------------------------------------------------
 1 file changed, 62 deletions(-)

diff --git a/src/filemanager/boxes.c b/src/filemanager/boxes.c
index 85cb18d..e823f50 100644
--- a/src/filemanager/boxes.c
+++ b/src/filemanager/boxes.c
@@ -102,7 +102,6 @@ static const int panel_listing_brief_idx = 1;
 static const int panel_listing_user_idx = 3;
 
 static char **status_format;
-static int listing_user_hotkey = 'u';
 static unsigned long panel_listing_types_id, panel_user_format_id, panel_brief_cols_id;
 static unsigned long mini_user_status_id, mini_user_format_id;
 
@@ -253,61 +252,6 @@ panel_listing_callback (Widget * w, Widget * sender, widget_msg_t msg, int parm,
 
     switch (msg)
     {
-    case MSG_KEY:
-        if (parm == '\n')
-        {
-            Widget *wi;
-
-            wi = dlg_find_by_id (h, panel_listing_types_id);
-            if (widget_is_active (wi))
-            {
-                WInput *in;
-
-                in = INPUT (dlg_find_by_id (h, mini_user_format_id));
-                input_assign_text (in, status_format[RADIO (wi)->sel]);
-                dlg_stop (h);
-                return MSG_HANDLED;
-            }
-
-            wi = dlg_find_by_id (h, panel_user_format_id);
-            if (widget_is_active (wi))
-            {
-                h->ret_value = B_USER + 6;
-                dlg_stop (h);
-                return MSG_HANDLED;
-            }
-
-            wi = dlg_find_by_id (h, mini_user_format_id);
-            if (widget_is_active (wi))
-            {
-                h->ret_value = B_USER + 7;
-                dlg_stop (h);
-                return MSG_HANDLED;
-            }
-        }
-
-        if (g_ascii_tolower (parm) == listing_user_hotkey)
-        {
-            Widget *wi;
-
-            wi = dlg_find_by_id (h, panel_user_format_id);
-            if (widget_is_active (wi))
-            {
-                wi = dlg_find_by_id (h, mini_user_format_id);
-                if (widget_is_active (wi))
-                {
-                    WRadio *r;
-
-                    r = RADIO (dlg_find_by_id (h, panel_listing_types_id));
-                    r->pos = r->sel = panel_listing_user_idx;
-                    dlg_select_widget (WIDGET (r));     /* force redraw */
-                    send_message (h, r, MSG_ACTION, 0, NULL);
-                    return MSG_HANDLED;
-                }
-            }
-        }
-        return MSG_NOT_HANDLED;
-
     case MSG_ACTION:
         if (sender != NULL && sender->id == panel_listing_types_id)
         {
@@ -773,7 +717,6 @@ panel_listing_box (WPanel * panel, int num, char **userp, char **minip, int *use
         char *panel_brief_cols_out = NULL;
         char *panel_user_format = NULL;
         char *mini_user_format = NULL;
-        const char *cp;
 
         /* Controls whether the array strings have been translated */
         const char *list_types[LIST_TYPES] = {
@@ -810,11 +753,6 @@ panel_listing_box (WPanel * panel, int num, char **userp, char **minip, int *use
             quick_widgets, panel_listing_callback, NULL
         };
 
-        /* get hotkey of user-defined format string */
-        cp = strchr (_(list_types[panel_listing_user_idx]), '&');
-        if (cp != NULL && *++cp != '\0')
-            listing_user_hotkey = g_ascii_tolower (*cp);
-
         mini_user_status = panel->user_mini_status;
         result = panel->list_type;
         status_format = panel->user_status_format;
-- 
2.5.0

