diff --git a/src/subshell/common.c b/src/subshell/common.c
index 1c432fd7e..90ec8d750 100644
--- a/src/subshell/common.c
+++ b/src/subshell/common.c
@@ -1778,12 +1778,16 @@ do_subshell_chdir (const vfs_path_t * vpath, gboolean update_prompt)
         }
     }
 
-    /* Really escape Zsh history */
-    if (mc_global.shell->type == SHELL_ZSH)
+    /* Really escape Zsh/Fish history */
+    if (mc_global.shell->type == SHELL_ZSH || mc_global.shell->type == SHELL_FISH)
     {
         /* Per Zsh documentation last command prefixed with space lingers in the internal history
          * until the next command is entered before it vanishes. To make it vanish right away,
-         * type a space and press return. */
+         * type a space and press return.
+         *
+         * Fish shell now also provides the same behavior:
+         * https://github.com/fish-shell/fish-shell/commit/9fdc4f903b8b421b18389a0f290d72cc88c128bb
+         * */
         write_all (mc_global.tty.subshell_pty, " \n", 2);
         subshell_state = RUNNING_COMMAND;
         feed_subshell (QUIETLY, TRUE);
