From 848ce6540ee08b51a05905d926574879266f39db Mon Sep 17 00:00:00 2001
From: Andreas Mohr <and@gmx.li>
Date: Sun, 20 Dec 2015 22:58:23 +0000
Subject: [PATCH] cleanup function declare at USE_DIFF_VIEW

Signed-off-by: Andreas Mohr <and@gmx.li>
---
 src/filemanager/cmd.h | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/src/filemanager/cmd.h b/src/filemanager/cmd.h
index cc57e49..4c62d9e 100644
--- a/src/filemanager/cmd.h
+++ b/src/filemanager/cmd.h
@@ -73,7 +73,9 @@ void edit_mc_menu_cmd (void);
 void edit_fhl_cmd (void);
 void hotlist_cmd (void);
 void compare_dirs_cmd (void);
+#ifdef USE_DIFF_VIEW
 void diff_view_cmd (void);
+#endif
 void panel_tree_cmd (void);
 void link_cmd (link_type_t link_type);
 void edit_symlink_cmd (void);
