From 18a44ea855f95e568c4702d92db0c19690c06226 Mon Sep 17 00:00:00 2001
From: Warlock <internalmike@gmail.com>
Date: Sat, 22 Apr 2017 20:58:34 +0300
Subject: [PATCH] group sticky bit fix

---
 src/filemanager/file.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/src/filemanager/file.c b/src/filemanager/file.c
index b61821a86..8e96e5584 100644
--- a/src/filemanager/file.c
+++ b/src/filemanager/file.c
@@ -2046,6 +2046,7 @@ copy_dir_dir (file_op_total_context_t * tctx, file_op_context_t * ctx, const cha
     struct link *lp;
     vfs_path_t *src_vpath, *dst_vpath;
     gboolean do_mkdir = TRUE;
+    struct stat temp_stat;
 
     src_vpath = vfs_path_from_str (s);
     dst_vpath = vfs_path_from_str (d);
@@ -2274,11 +2275,12 @@ copy_dir_dir (file_op_total_context_t * tctx, file_op_context_t * ctx, const cha
     }
     mc_closedir (reading);
 
+    mc_stat(dst_vpath, &temp_stat);
     if (ctx->preserve)
     {
         mc_timesbuf_t times;
 
-        mc_chmod (dst_vpath, cbuf.st_mode & ctx->umask_kill);
+        mc_chmod (dst_vpath, (cbuf.st_mode & ctx->umask_kill) | (temp_stat.st_mode & 007000));
         get_times (&cbuf, &times);
         mc_utime (dst_vpath, &times);
     }
@@ -2287,7 +2289,7 @@ copy_dir_dir (file_op_total_context_t * tctx, file_op_context_t * ctx, const cha
         cbuf.st_mode = umask (-1);
         umask (cbuf.st_mode);
         cbuf.st_mode = 0100777 & ~cbuf.st_mode;
-        mc_chmod (dst_vpath, cbuf.st_mode & ctx->umask_kill);
+        mc_chmod (dst_vpath, (cbuf.st_mode & ctx->umask_kill) | (temp_stat.st_mode & 007000));
     }
 
   ret:
-- 
2.11.0

