diff --git a/misc/mc.menu.in b/misc/mc.menu.in
index 2bf39c48e..d7b3dd511 100644
--- a/misc/mc.menu.in
+++ b/misc/mc.menu.in
@@ -228,7 +228,7 @@ U       Uudecode marked news articles (needs work)
         fi
         echo "Please test the output file before deleting anything."
 
-=+ f \.tar\.gz$ | f \.tar\.z$ | f \.tgz$ | f \.tpz$ | f \.tar\.lz$ | f \.tar\.lz4$ | f \.tar\.lzma$ | f \.tar\.7z$ | f \.tar\.xz$ | f \.tar\.zst | f \.tar\.Z$ | f \.tar\.bz2$ & t r
+=+ f \.tar\.gz$ | f \.tar\.z$ | f \.tgz$ | f \.tpz$ | f \.tar\.lz$ | f \.tar\.lz4$ | f \.tar\.lzma$ | f \.tar\.7z$ | f \.tar\.xz$ | f \.tar\.zst | f \.tar\.Z$ | f \.tar\.bz2$ & t rl
 x       Extract the contents of a compressed tar file
         unset PRG
         case %f in
