diff -urN mc-4.6.2_mcDC/ChangeLog mc-2009-04-05/ChangeLog
--- mc-4.6.2_MCDC/ChangeLog	2009-02-01 19:30:21.000000000 +0000
+++ mc-2009-04-05/ChangeLog	2009-04-05 12:09:30.000000000 +0100
@@ -1,3 +1,11 @@
+2009-04-05  Vit Rosin   <vit_r@list.ru>
+
+	* edit.c, edit.h : Add edit_an_ordinary_key_press ()
+	Usually after typing comma-key (prints comma char ',')
+	spacebar press follows to insert space ' '.
+	With this patch typing comma-key prints two chars
+	comma char ',' and space ' ' (except some cases).
+
 2007-11-02  Vladimir Nadvornik  <nadvornik@suse.cz>
 
 	* editlock.c (lock_build_name): Check the return value of getpwuid().
diff -urN mc-4.6.2_MCDC/edit.c mc-2009-04-05/edit.c
--- mc-4.6.2_MCDC/edit.c	2009-02-01 19:30:21.000000000 +0000
+++ mc-2009-04-05/edit.c	2009-04-05 12:09:30.000000000 +0100
@@ -2176,7 +2176,7 @@
 	    if (edit_get_byte (edit, edit->curs1) != '\n')
 		edit_delete (edit);
 	}
-	edit_insert (edit, char_for_insertion);
+        edit_an_ordinary_key_press (edit, char_for_insertion);
 	if (option_auto_para_formatting) {
 	    format_paragraph (edit, 0);
 	    edit->force |= REDRAW_PAGE;
@@ -2701,3 +2701,33 @@
 cleanup:
     g_free (block_file);
 }
+
+void edit_an_ordinary_key_press (WEdit * edit, const int char_for_insertion)
+{
+    int c0, c2;
+    if (',' == char_for_insertion) {
+        edit_insert (edit, char_for_insertion);
+        c0 = edit_get_byte (edit, edit->curs1);
+        if (isdigit (c0) || '\'' == c0)
+            return;
+
+        if (edit->curs1 > 1) {
+            c2 = edit_get_byte (edit, edit->curs1 - 2);
+            if (isdigit (c2) || '\'' == c2)
+                return;
+        }
+        if ( ! (' ' == c0 || '\t' == c0 || '\'' == c0))
+            edit_insert (edit, ' ');
+
+        return;
+
+    } else if (isdigit (char_for_insertion)) {
+        if (edit->curs1 > 2) {
+            if (' ' == edit_get_byte (edit, edit->curs1 - 1)
+                    && ',' == edit_get_byte (edit, edit->curs1 - 2)
+                    && isdigit (edit_get_byte (edit, edit->curs1 - 3)))
+                edit_backspace (edit);
+        }
+    }
+    edit_insert (edit, char_for_insertion);
+}
diff -urN mc-4.6.2_MCDC/edit.h mc-2009-04-05/edit.h
--- mc-4.6.2_MCDC/edit.h	2009-02-01 19:30:21.000000000 +0000
+++ mc-2009-04-05/edit.h	2009-04-05 12:09:30.000000000 +0100
@@ -234,6 +234,7 @@
 
 /* either command or char_for_insertion must be passed as -1 */
 void edit_execute_cmd (WEdit *edit, int command, int char_for_insertion);
+void edit_an_ordinary_key_press (WEdit * edit, const int char_for_insertion);
 
 #define get_sys_error(s) (s)
 
