diff -Naur mc-4.7.0-pre1~/src/search/normal.c mc-4.7.0-pre1/src/search/normal.c
--- mc-4.7.0-pre1~/src/search/normal.c	2009-06-23 21:55:12.000000000 +0300
+++ mc-4.7.0-pre1/src/search/normal.c	2009-08-09 23:02:09.000000000 +0300
@@ -68,6 +68,7 @@
         case ')':
         case '^':
         case '-':
+        case '|':
             g_string_append_c (buff, '\\');
             g_string_append_c (buff, str[loop]);
             loop++;
