--- mc/src/vfs/extfs/helpers/iso9660.in.orig	2014-03-08 03:04:44.000000000 +0100
+++ mc/src/vfs/extfs/helpers/iso9660.in	2014-03-08 03:08:00.000000000 +0100
@@ -37,14 +37,17 @@
     if test -z "$XORRISO"; then
         return 1
     fi
-    local dir attr ln usr gr sz dt1 dt2 dt3 nm len name
-    dir="$2"
-    $XORRISO -dev stdio:"$1" -cd "$dir" -lsl 2> /dev/null | @GREP@ "^[-d]" | \
+    local dir attr ln usr gr sz dt1 dt2 dt3 nm len name lsl r
+    dir="${2:-/}"
+    lsl=$( $XORRISO -abort_on FATAL -dev stdio:"$1" -cd "$dir" -lsl 2> /dev/null )
+    r=$?
+    test $r -gt 0 && return $r 
+ 	echo "$lsl" | @GREP@ "^[-d]" | \
     while read attr ln usr gr sz dt1 dt2 dt3 nm ; do
         len=$((${#nm} - 1))
         name=$(printf "$nm" | cut -c2-$len)  # remove quotes
         if test $(printf "$nm" | cut -c1-2) != "d"; then
-            printf "%s %s %s %s %s %s %s %s %s/%s\n" "$attr" "$ln" "$usr" "$gr" "$sz" "$dt1" "$dt2" "$dt3" "$dir" "$name"
+            printf -- "%s %s %s %s %s %s %s %s %s/%s\n" "$attr" "$ln" "$usr" "$gr" "$sz" "$dt1" "$dt2" "$dt3" "$dir" "$name"
         else
             xorriso_list "$1" "$dir/$name"
         fi
