From 85c657402d5618bcc1e21ffbb45f6c31069060bc Mon Sep 17 00:00:00 2001
From: "Yury V. Zaytsev" <yury@shurup.com>
Date: Thu, 17 Oct 2024 09:15:29 +0200
Subject: [PATCH] buildsys: fix bootstrapped build on Solaris and systems with
 LC_ALL != C

Signed-off-by: Yury V. Zaytsev <yury@shurup.com>
---
 autogen.sh | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/autogen.sh b/autogen.sh
index f17c51a9d..94d8e7c3f 100755
--- a/autogen.sh
+++ b/autogen.sh
@@ -14,8 +14,9 @@ ${AUTORECONF:-autoreconf} --verbose --install --force -I m4 ${AUTORECONF_FLAGS}
 rm -f INSTALL && ln -s doc/INSTALL .
 
 # Generate po/POTFILES.in
-if ! xgettext -h 2>&1 | grep -e '--keyword=' >/dev/null ; then
+if ! xgettext -h 2>&1 | grep -e '--keyword' >/dev/null ; then
   echo "gettext is unable to extract translations, set XGETTEXT to GNU gettext!" >&2
+  touch po/POTFILES.in
 else
   ${XGETTEXT:-xgettext} --keyword=_ --keyword=N_ --keyword=Q_ --output=- \
 	`find . -name '*.[ch]'` | ${SED-sed} -ne '/^#:/{s/#://;s/:[0-9]*/\
-- 
2.39.5 (Apple Git-154)

