Fix ticket #3698 - Fix mcedit codejump (etags filenames parsing)

Signed-off-by: Roman Mikhayloff <rimf@inbox.ru>

diff --git a/src/editor/etags.c b/src/editor/etags.c
index 4f14169..9bdb0d8 100644
--- a/src/editor/etags.c
+++ b/src/editor/etags.c
@@ -214,7 +214,7 @@ etags_set_definition_hash (const char *tagfile, const char *start_path,
 
                 pos = strcspn (buf, ",");
                 g_free (filename);
-                filename = g_strndup (buf, pos + 1);
+                filename = g_strndup (buf, pos);
                 state = in_define;
                 break;
             }
-- 
2.10.0

