diff --git a/src/viewer/dialogs.c b/src/viewer/dialogs.c
index b93fc18..44521cb 100644
--- a/src/viewer/dialogs.c
+++ b/src/viewer/dialogs.c
@@ -238,6 +238,8 @@ mcview_dialog_goto (mcview_t * view, off_t * offset)
             switch (current_goto_type)
             {
             case MC_VIEW_GOTO_LINENUM:
+                if (addr > 0)
+                    addr--;
                 mcview_coord_to_offset (view, offset, addr, 0);
                 *offset = mcview_bol (view, *offset, 0);
                 break;
