--- mc.git/src/args.c	2009-09-02 20:22:20.000000000 +0200
+++ mc.src/src/args.c	2009-09-02 20:53:10.000000000 +0200
@@ -283,11 +283,10 @@
      * FIXME: undocumented keywords: viewunderline, editnormal, editbold,
      * and editmarked.  To preserve translations, lines should be split.
      */
-    /* TRANSLATORS: don't translate keywords and names of colors */
 
     mc_args__loc__colors_string = g_strdup_printf("%s%s",
-	_
-	   ("--colors KEYWORD={FORE},{BACK}\n\n"
+	/* TRANSLATORS: don't translate keywords and names of colors */
+	_(  "--colors KEYWORD={FORE},{BACK}\n\n"
 	    "{FORE} and {BACK} can be omitted, and the default will be used\n"
 	    "\n" "Keywords:\n"
 	    "   Global:       errors, reverse, gauge, input, viewunderline\n"
@@ -297,9 +296,8 @@
 	    "   Menus:        menu, menuhot, menusel, menuhotsel\n"
 	    "   Editor:       editnormal, editbold, editmarked, editwhitespace,\n"
 	    "                 editlinestate\n"),
-	_
-	   (
-	    "   Help:         helpnormal, helpitalic, helpbold, helplink, helpslink\n"
+	/* TRANSLATORS: don't translate keywords and names of colors */
+	_(  "   Help:         helpnormal, helpitalic, helpbold, helplink, helpslink\n"
 	    "   File types:   directory, executable, link, stalelink, device, special, core\n"
 	    "\n" "Colors:\n"
 	    "   black, gray, red, brightred, green, brightgreen, brown,\n"
