diff --git a/lib/widget/listbox.c b/lib/widget/listbox.c
index f411851..0d94a83 100644
--- a/lib/widget/listbox.c
+++ b/lib/widget/listbox.c
@@ -352,8 +352,10 @@ listbox_key (WListbox * l, int key)
 static inline void
 listbox_append_item (WListbox * l, WLEntry * e, listbox_append_t pos)
 {
-    if (l->list == NULL)
+    if (l->list == NULL) {
         l->list = g_queue_new ();
+        pos = LISTBOX_APPEND_AT_END;
+    }
 
     switch (pos)
     {
