diff --git a/misc/skins/nicedark.ini b/misc/skins/nicedark.ini
index bccdb19..a86b025 100644
--- a/misc/skins/nicedark.ini
+++ b/misc/skins/nicedark.ini
@@ -33,7 +33,7 @@
     input=lightgray;black
     inputunchanged=gray;black
     inputmark=black;lightgray
-    disabled=gray;darkgray
+    disabled=gray;black
     reverse=brown;black
     commandlinemark=black;lightgray
     header=lightgray;black
@@ -41,11 +41,11 @@
     commandhistory=
 
 [dialog]
-    _default_=lightgray;darkgray
+    _default_=lightgray;black
     dfocus=lightgray;blue
-    dhotnormal=brown;darkgray
+    dhotnormal=brown;black
     dhotfocus=brown;blue
-    dtitle=brown;darkgray
+    dtitle=brown;black
 
 [error]
     _default_=red;black
@@ -86,15 +86,15 @@
     _default_=lightgray;black
 
 [help]
-    _default_=lightgray;darkgray
-    helpitalic=red;darkgray
-    helpbold=brown;darkgray
-    helplink=blue;darkgray
+    _default_=lightgray;black
+    helpitalic=red;black
+    helpbold=brown;black
+    helplink=blue;black
     helpslink=lightgray;blue
-    helptitle=brown;darkgray
+    helptitle=brown;black
 
 [editor]
-    _default_=lightgray;darkgray
+    _default_=lightgray;black
     editbold=brown;blue
     editmarked=black;cyan
     editwhitespace=brightblue;blue
