diff -ur mc-4.8.0.orig/misc/skins/sand256.ini mc-4.8.0/misc/skins/sand256.ini
--- mc-4.8.0.orig/misc/skins/sand256.ini	2011-11-07 21:27:32.000000000 +0100
+++ mc-4.8.0/misc/skins/sand256.ini	2011-11-07 22:11:29.000000000 +0100
@@ -97,11 +97,14 @@
     dfocus=;rgb452
     dhotnormal=;;underline
     dhotfocus=;rgb452;underline
+    dtitle=;;underline
 
 [error]
     _default_=rgb554;rgb320;bold
+    errdfocus=black;rgb452;bold
     errdhotnormal=;;bold+underline
-    errdhotfocus=black;rgb452;underline
+    errdhotfocus=black;rgb452;bold+underline
+    errdtitle=;;bold+underline
 
 [filehighlight]
     directory=
