diff --git a/src/viewer/inlines.h b/src/viewer/inlines.h
index 224f8e6..6538720 100644
--- a/src/viewer/inlines.h
+++ b/src/viewer/inlines.h
@@ -16,7 +16,7 @@
 static inline off_t
 mcview_offset_doz (off_t a, off_t b)
 {
-    return (a >= b) ? a - b : 0;
+    return (b < a) ? (a - b) : 0;
 }
 
 /* --------------------------------------------------------------------------------------------- */
@@ -34,7 +34,7 @@ mcview_offset_rounddown (off_t a, off_t b)
 static inline screen_dimen
 mcview_dimen_doz (screen_dimen a, screen_dimen b)
 {
-    return (a >= b) ? a - b : 0;
+    return (b < a) ? a - b : 0;
 }
 
 /* --------------------------------------------------------------------------------------------- */
