diff --git a/doc/man/mcedit.1.in b/doc/man/mcedit.1.in
index e33641db8..b27f8d471 100644
--- a/doc/man/mcedit.1.in
+++ b/doc/man/mcedit.1.in
@@ -604,39 +604,6 @@ will be untouched. Default value is
 Show full path name in the status line. If disabled (default), only base name of the
 file is shown.
 .SH MISCELLANEOUS
-You can use scanf search and replace to search and replace a C format
-string.  First take a look at the
-.B sscanf
-and
-.B sprintf
-man pages to see what a format string is and how it works.  Here's an
-example: suppose that you want to replace all occurrences of an open
-bracket, three comma separated numbers, and a close bracket, with the
-word
-.IR apples ,
-the third number, the word
-.I oranges
-and then the second number.  You would fill in the Replace dialog box as
-follows:
-.PP
-.nf
-.B Enter search string
-(%d,%d,%d)
-.B Enter replace string
-apples %d oranges %d
-.B Enter replacement argument order
-3,2
-.fi
-.PP
-The last line specifies that the third and then the second number are to
-be used in place of the first and second.
-.PP
-It is advisable to use this feature with Prompt On Replace on, because a
-match is thought to be found whenever the number of arguments found
-matches the number given, which is not always a real match. Scanf also
-treats whitespace as being elastic.  Note that the scanf format %[ is
-very useful for scanning strings, and whitespace.
-.PP
 The editor also displays non\-us characters (160+).  When editing
 binary files, you should set
 .B display bits
