--- mc.git/src/args.c	2017-04-16 20:32:47.761872908 +0200
+++ mc.src/src/args.c	2017-11-26 21:58:22.568006145 +0100
@@ -137,7 +137,7 @@
      "printwd", 'P', G_OPTION_FLAG_IN_MAIN, G_OPTION_ARG_STRING,
      &mc_args__last_wd_file,
      N_("Print last working directory to specified file"),
-     "<file>"
+     N_("<file>")
     },
 
 #ifdef ENABLE_SUBSHELL
@@ -162,7 +162,7 @@
      "ftplog", 'l', G_OPTION_FLAG_IN_MAIN, G_OPTION_ARG_STRING,
      &mc_args__netfs_logfile,
      N_("Log ftp dialog to specified file"),
-     "<file>"
+     N_("<file>")
     },
 #endif /* ENABLE_VFS_FTP */
 #ifdef ENABLE_VFS_SMB
@@ -170,7 +170,7 @@
      "debuglevel", 'D', G_OPTION_FLAG_IN_MAIN, G_OPTION_ARG_INT,
      &mc_args__debug_level,
      N_("Set debug level"),
-     "<integer>"
+     N_("<integer>")
     },
 #endif /* ENABLE_VFS_SMB */
 
@@ -179,7 +179,7 @@
      "view", 'v', G_OPTION_FLAG_IN_MAIN | G_OPTION_FLAG_NO_ARG, G_OPTION_ARG_CALLBACK,
      (gpointer) parse_mc_v_argument,
      N_("Launches the file viewer on a file"),
-     "<file>"
+     N_("<file>")
     },
 
     {
@@ -187,7 +187,7 @@
      "edit", 'e', G_OPTION_FLAG_IN_MAIN | G_OPTION_FLAG_NO_ARG, G_OPTION_ARG_CALLBACK,
      (gpointer) parse_mc_e_argument,
      N_("Edit files"),
-     "<file> ..." },
+     N_("<file> ...") },
 
     {
      NULL, '\0', 0, 0, NULL, NULL, NULL /* Complete struct initialization */
@@ -264,7 +264,7 @@
      "keymap", 'K', ARGS_TERM_OPTIONS, G_OPTION_ARG_STRING,
      &mc_args__keymap_file,
      N_("Load definitions of key bindings from specified file"),
-     "<file>"
+     N_("<file>")
     },
 
     {
@@ -306,14 +306,14 @@
      "colors", 'C', ARGS_COLOR_OPTIONS, G_OPTION_ARG_STRING,
      &mc_global.tty.command_line_colors,
      N_("Specifies a color configuration"),
-     "<string>"
+     N_("<string>")
     },
 
     {
      "skin", 'S', ARGS_COLOR_OPTIONS, G_OPTION_ARG_STRING,
      &mc_global.tty.skin,
      N_("Show mc with specified skin"),
-     "<string>"
+     N_("<string>")
     },
 
     {
