diff --git a/lib/widget/wtools.c b/lib/widget/wtools.c
index 9054d7c..abbb576 100644
--- a/lib/widget/wtools.c
+++ b/lib/widget/wtools.c
@@ -372,11 +372,13 @@ message (int flags, const char *title, const char *text, ...)
 {
     char *p;
     va_list ap;
+#ifdef WITH_BACKGROUND
     union
     {
         void *p;
         void (*f) (int, int *, char *, const char *);
     } func;
+#endif /* WITH_BACKGROUND */
 
     va_start (ap, text);
     p = g_strdup_vprintf (text, ap);
@@ -410,12 +412,12 @@ char *
 input_dialog_help (const char *header, const char *text, const char *help,
                    const char *history_name, const char *def_text)
 {
+#ifdef WITH_BACKGROUND
     union
     {
         void *p;
         char *(*f) (const char *, const char *, const char *, const char *, const char *);
     } func;
-#ifdef WITH_BACKGROUND
     if (we_are_background)
     {
         func.f = fg_input_dialog_help;
