diff -ur mc-4.8.0.orig/src/filemanager/file.c mc-4.8.0/src/filemanager/file.c
--- mc-4.8.0.orig/src/filemanager/file.c	2011-10-18 13:39:19.000000000 +0200
+++ mc-4.8.0/src/filemanager/file.c	2011-11-21 19:03:49.908638355 +0100
@@ -1522,18 +1522,20 @@
         struct timeval tv_current, tv_last_update, tv_last_input;
         int secs, update_secs;
         const char *stalled_msg = "";
+        int read_size;
+        char buf[BUF_8K];
 
         tv_last_update = tv_transfer_start;
 
+        read_size = sizeof (buf) - ctx->do_reget % sizeof (buf);
+
         for (;;)
         {
-            char buf[BUF_8K];
-
             /* src_read */
             if (mc_ctl (src_desc, VFS_CTL_IS_NOTREADY, 0))
                 n_read = -1;
             else
-                while ((n_read = mc_read (src_desc, buf, sizeof (buf))) < 0 && !ctx->skip_all)
+                while ((n_read = mc_read (src_desc, buf, read_size)) < 0 && !ctx->skip_all)
                 {
                     return_status = file_error (_("Cannot read source file\"%s\"\n%s"), src_path);
                     if (return_status == FILE_RETRY)
@@ -1545,6 +1547,8 @@
             if (n_read == 0)
                 break;
 
+            read_size = sizeof (buf);
+
             gettimeofday (&tv_current, NULL);
 
             if (n_read > 0)
