--- file.c1	2010-04-10 01:53:25.000000000 +0400
+++ file.c	2010-04-26 00:50:22.000000000 +0400
@@ -1956,7 +1956,16 @@
 		&& dest_dir[0] != '\0'
 		&& dest_dir[strlen (dest_dir) - 1] != PATH_SEP) {
 	    /* add trailing separator */
+	    if (single_entry) { // nickk: add filename to dest path
+	        dest_dir_ = g_strconcat (dest_dir, PATH_SEP_STR, source, (char *) NULL);
+	    } else {
 	    dest_dir_ = g_strconcat (dest_dir, PATH_SEP_STR, (char *) NULL);
+    	    }
+    	} else if (!force_single && single_entry
+		&& dest_dir[0] != '\0'
+		&& dest_dir[strlen (dest_dir) - 1] == PATH_SEP) {
+	    // nickk: add filename to dest path
+            dest_dir_ = g_strconcat (dest_dir, source, (char *) NULL);
 	} else {
 	    /* just copy */
 	    dest_dir_ = g_strdup (dest_dir);
