From 2ffd5c50fba962cad6e3f713cdc51267c69ef74c Mon Sep 17 00:00:00 2001
From: Vit Rosin <vit_r@list.ru>
Date: Tue, 17 Nov 2009 15:47:38 +0000
Subject: [PATCH] src_keybindC_mistakes2

---
 src/keybind.c |    1 -
 1 files changed, 0 insertions(+), 1 deletions(-)

diff --git a/src/keybind.c b/src/keybind.c
index c898e8c..652da4e 100644
--- a/src/keybind.c
+++ b/src/keybind.c
@@ -576,7 +576,6 @@ const global_keymap_t default_editor_keymap[] = {
     { KEY_M_ALT | KEY_END,                  CK_End_Alt_Highlight,           "M-End" },
 
     /* Ctrl */
-    { KEY_M_CTRL | (KEY_F (2)),             CK_Save_As,                     "C-F2" },
     { KEY_M_CTRL | (KEY_F (4)),             CK_Replace_Again,               "C-F4" },
     { KEY_M_CTRL | (KEY_F (7)),             CK_Find_Again,                  "C-F7" },
     { KEY_M_CTRL | KEY_BACKSPACE,           CK_Undo,                        "C-BackSpase" },
-- 
1.6.4

