From 6e49741528ff56e9e722a3ea6afd6c708ecef5ff Mon Sep 17 00:00:00 2001
From: Mooffie <mooffie@gmail.com>
Date: Mon, 19 Dec 2016 02:04:10 +0200
Subject: [PATCH 1/6] Ticket #2707: extfs: u7z: make the code more readable.

---
 src/vfs/extfs/helpers/u7z | 18 ++++++++++++++----
 1 file changed, 14 insertions(+), 4 deletions(-)

diff --git a/src/vfs/extfs/helpers/u7z b/src/vfs/extfs/helpers/u7z
index 22e9f27..8270a20 100755
--- a/src/vfs/extfs/helpers/u7z
+++ b/src/vfs/extfs/helpers/u7z
@@ -23,18 +23,28 @@ mcu7zip_list ()
 {
         # Symlinks are not shown - no idea how to distinguish them
         # Read-only files are not shown as such - it's rarely useful
+
         ugid="`id -nu` `id -ng`"
+
         date_re='^\(....\)-\(..\)-\(..\) \(..:..:..\)'
         date_mc='\2-\3-\1 \4'
         size_re='............'
+
         # archive entries can have no datetime info, 7z will use archive file datetime
         date_archive=`stat -c %y "$1" 2>/dev/null | sed -n "s/${date_re}.*/${date_mc}/p" 2>/dev/null`
         [ "${date_archive}"x = x ] && date_archive=`ls -lan "$1" 2>/dev/null | awk '{print $6,"",$7,"",$8}' 2>/dev/null`
         [ "${date_archive}"x = x ] && date_archive="01-01-1970 00:00:00"
-        $P7ZIP l "$1" | sed -n "s/$date_re D.... $size_re $size_re\(.*\)/drwxr-xr-x 1 $ugid 0 $date_mc \5/p;\
-                                s/$date_re \..... \($size_re\) $size_re\(.*\)/-rw-r--r-- 1 $ugid \5 $date_mc \6/p;\
-                                s/^\s*D.... $size_re $size_re\(.*\)/drwxr-xr-x 1 $ugid 0 $date_archive \1/p;\
-                                s/^\s*\..... \($size_re\) $size_re\(.*\)/-rw-r--r-- 1 $ugid \1 $date_archive \2/p"
+
+        $P7ZIP l "$1" | sed -n "
+
+                  # Handle directories.
+                  s/$date_re D.... $size_re $size_re\(.*\)/drwxr-xr-x 1 $ugid 0 $date_mc \5/p
+                       s/^\s*D.... $size_re $size_re\(.*\)/drwxr-xr-x 1 $ugid 0 $date_archive \1/p
+
+                  # Handle normal files.
+                  s/$date_re \..... \($size_re\) $size_re\(.*\)/-rw-r--r-- 1 $ugid \5 $date_mc \6/p
+                       s/^\s*\..... \($size_re\) $size_re\(.*\)/-rw-r--r-- 1 $ugid \1 $date_archive \2/p
+        "
 }
 
 mcu7zip_copyout ()
-- 
2.9.3

