From c6e08009488e636b04791b25a28e57748aa59797 Mon Sep 17 00:00:00 2001
From: Jindrich Makovicka <makovick@gmail.com>
Date: Mon, 8 Aug 2011 23:29:00 +0200
Subject: [PATCH] fix autoconf warning

Signed-off-by: Jindrich Makovicka <makovick@gmail.com>
---
 m4.include/mc-with-screen-slang.m4 |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/m4.include/mc-with-screen-slang.m4 b/m4.include/mc-with-screen-slang.m4
index 876b839..d1b4b28 100644
--- a/m4.include/mc-with-screen-slang.m4
+++ b/m4.include/mc-with-screen-slang.m4
@@ -78,7 +78,7 @@ AC_DEFUN([MC_CHECK_SLANG_BY_PATH], [
     dnl check if S-Lang have version 2.0 or newer
     if test x"$found_slang" = x"yes"; then
         AC_MSG_CHECKING([for S-Lang version 2.0 or newer])
-        AC_RUN_IFELSE([
+        AC_RUN_IFELSE([AC_LANG_SOURCE([
 #ifdef HAVE_SLANG_SLANG_H
 #include <slang/slang.h>
 #else
@@ -92,7 +92,7 @@ int main (void)
     return 1;
 #endif
 }
-],
+])],
 	    [mc_slang_is_valid_version=yes],
 	    [mc_slang_is_valid_version=no],
 	    [
-- 
1.7.5.4

