From 1ecce4642868d1030e27ad138a8d8a9415c86174 Mon Sep 17 00:00:00 2001
From: Vit Rosin <vit_r@list.ru>
Date: Mon, 8 Mar 2010 12:16:45 +0000
Subject: [PATCH]  repl deprecated g_tree_traverse by g_tree_foreach

---
 src/editor/syntax.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/src/editor/syntax.c b/src/editor/syntax.c
index 0f590dc..45a1bed 100644
--- a/src/editor/syntax.c
+++ b/src/editor/syntax.c
@@ -135,7 +135,7 @@ mc_defines_destroy (gpointer key, gpointer value, gpointer data)
 static void
 destroy_defines (GTree ** defines)
 {
-    g_tree_traverse (*defines, mc_defines_destroy, G_POST_ORDER, NULL);
+    g_tree_foreach (*defines, mc_defines_destroy, NULL);
     g_tree_destroy (*defines);
     *defines = NULL;
 }
-- 
1.6.4

