diff --git a/lib/mc.ext.in b/lib/mc.ext.in
index c81aa71..6335b69 100644
--- a/lib/mc.ext.in
+++ b/lib/mc.ext.in
@@ -446,7 +446,7 @@ type/^PDF
 # html
 regex/\.([hH][tT][mM][lL]?)$
 	Open=(if test -n "@X11_WWW@" && test -n "$DISPLAY"; then (@X11_WWW@ file://%d/%p &) 1>&2; else links %f || lynx -force_html %f || ${PAGER:-more} %f; fi) 2>/dev/null
-	View=%view{ascii} lynx -dump -force_html %f
+	View=%view{ascii} links -dump %f 2>/dev/null || w3m -dump %f 2>/dev/null || lynx -dump -force_html %f
 
 # StarOffice 5.2
 shell/.sdw
