diff --git a/misc/ext.d/doc.sh.in b/misc/ext.d/doc.sh.in
index c296ca270..8baec0754 100644
--- a/misc/ext.d/doc.sh.in
+++ b/misc/ext.d/doc.sh.in
@@ -52,7 +52,10 @@ do_view_action() {
         fi
         ;;
     msdoc)
-        if which wvHtml >/dev/null 2>&1; then
+        OOFFICE=`get_ooffice_executable`
+        if [ -n "${OOFFICE}" ]; then
+            ${OOFFICE} --headless --cat "${MC_EXT_FILENAME}"
+        elif which wvHtml >/dev/null 2>&1; then
             tmp=`mktemp -d ${TMPDIR:-/tmp}/%p.XXXXXX`
             wvHtml "${MC_EXT_FILENAME}" --targetdir="$tmp" page.html
             elinks -dump "$tmp/page.html"
