diff -r -u -w mc-4.7.1-org/src/cmd.c mc-4.7.1-viewertoeditor/src/cmd.c
--- mc-4.7.1-org/src/cmd.c	2010-02-26 14:34:03.000000000 +0300
+++ mc-4.7.1-viewertoeditor/src/cmd.c	2010-04-26 02:12:58.000000000 +0400
@@ -117,6 +117,7 @@
 {
     static const char *viewer = NULL;
     int move_dir = 0;
+    int run_editor = 0;
 
 
     if (plain_view) {
@@ -136,7 +137,7 @@
 	mcview_default_hex_mode = 0;
 	mcview_default_nroff_flag = 0;
 	mcview_default_magic_flag = 0;
-	mcview_viewer (NULL, filename, &move_dir, start_line);
+	mcview_viewer (NULL, filename, &move_dir, start_line, &run_editor);
 	if (changed_hex_mode && !mcview_altered_hex_mode)
 	    mcview_default_hex_mode = 1;
 	if (changed_nroff_flag && !mcview_altered_nroff_flag)
@@ -144,6 +145,10 @@
 	if (changed_magic_flag && !mcview_altered_magic_flag)
 	    mcview_default_magic_flag = 1;
 	repaint_screen ();
+        if (run_editor) {
+		do_edit_at_line (filename, start_line);
+		repaint_screen ();
+        }
 	return move_dir;
     }
     if (internal) {
@@ -156,7 +161,7 @@
 	    strcpy (view_entry, "View");
 
 	if (regex_command (filename, view_entry, &move_dir) == 0) {
-	    mcview_viewer (NULL, filename, &move_dir, start_line);
+	    mcview_viewer (NULL, filename, &move_dir, start_line, &run_editor);
 	    repaint_screen ();
 	}
     } else {
@@ -169,6 +174,10 @@
 	}
 	execute_with_vfs_arg (viewer, filename);
     }
+    if (run_editor) {
+	do_edit_at_line (filename, start_line);
+	repaint_screen ();
+    }
     return move_dir;
 }
 
@@ -289,7 +298,7 @@
     if (!command)
 	return;
 
-    mcview_viewer (command, "", NULL, 0);
+    mcview_viewer (command, "", NULL, 0, NULL);
 
     g_free (command);
 }
diff -r -u -w mc-4.7.1-org/src/cmddef.h mc-4.7.1-viewertoeditor/src/cmddef.h
--- mc-4.7.1-org/src/cmddef.h	2010-02-26 14:34:03.000000000 +0300
+++ mc-4.7.1-viewertoeditor/src/cmddef.h	2010-04-26 02:16:42.000000000 +0400
@@ -274,6 +274,7 @@
 #define CK_ViewPrevFile                 5025
 #define CK_ViewToggleRuler              5026
 #define CK_ViewToggleHexNavMode         5027
+#define CK_ViewRunEditor                5028
 
 /* Tree */
 #define CK_TreeHelp                     6001
 
diff -r -u -w mc-4.7.1-org/src/ext.c mc-4.7.1-viewertoeditor/src/ext.c
--- mc-4.7.1-org/src/ext.c	2010-02-26 14:34:03.000000000 +0300
+++ mc-4.7.1-viewertoeditor/src/ext.c	2010-04-26 02:14:59.000000000 +0400
@@ -252,10 +252,10 @@
 	 * into view
 	 */
 	if (written_nonspace) {
-	    mcview_viewer (cmd, filename, move_dir, start_line);
+	    mcview_viewer (cmd, filename, move_dir, start_line, NULL);
 	    unlink (file_name);
 	} else {
-	    mcview_viewer (NULL, filename, move_dir, start_line);
+	    mcview_viewer (NULL, filename, move_dir, start_line, NULL);
 	}
 	if (changed_hex_mode && !mcview_altered_hex_mode)
 	    mcview_default_hex_mode = def_hex_mode;
diff -r -u -w mc-4.7.1-org/src/keybind.c mc-4.7.1-viewertoeditor/src/keybind.c
--- mc-4.7.1-org/src/keybind.c	2010-02-26 14:34:03.000000000 +0300
+++ mc-4.7.1-viewertoeditor/src/keybind.c	2010-04-26 02:18:36.000000000 +0400
@@ -249,6 +249,7 @@
     { "ViewPrevFile",                    CK_ViewPrevFile },
     { "ViewToggleRuler",                 CK_ViewToggleRuler },
     { "ViewToggleHexNavMode",            CK_ViewToggleHexNavMode },
+    { "ViewRunEditor",                   CK_ViewRunEditor },
 
     /* help */
     { "HelpHelp",                        CK_HelpHelp },
@@ -514,6 +515,8 @@
     { ALT ('e'),   CK_SelectCodepage,        "M-e" },
     { XCTRL ('o'), CK_ShowCommandLine,       "C-o" },
 
+    { KEY_F (14),  CK_ViewRunEditor,        "S-F4" },
+
     { 0, CK_Ignore_Key, "" }
 };
 
diff -r -u -w mc-4.7.1-org/src/main.c mc-4.7.1-viewertoeditor/src/main.c
--- mc-4.7.1-org/src/main.c	2010-02-26 14:34:03.000000000 +0300
+++ mc-4.7.1-viewertoeditor/src/main.c	2010-04-26 02:01:07.000000000 +0400
@@ -1604,7 +1604,8 @@
 	}
 
 	/* Ctrl-Shift-Enter */
-	if (parm == (KEY_M_CTRL | KEY_M_SHIFT | '\n')) {
+	if (parm == (KEY_M_CTRL | KEY_M_SHIFT | '\n')) 
+	{
 	    copy_current_pathname ();
 	    copy_prog_name ();
 	    return MSG_HANDLED;
diff -r -u -w mc-4.7.1-org/src/user.c mc-4.7.1-viewertoeditor/src/user.c
--- mc-4.7.1-org/src/user.c	2010-02-26 14:34:03.000000000 +0300
+++ mc-4.7.1-viewertoeditor/src/user.c	2010-04-26 02:15:17.000000000 +0400
@@ -702,7 +702,7 @@
     chmod (file_name, S_IRWXU);
     if (run_view) {
 	run_view = 0;
-	mcview_viewer (file_name, NULL, &run_view, 0);
+	mcview_viewer (file_name, NULL, &run_view, 0, NULL);
     } else {
 	/* execute the command indirectly to allow execution even
 	 * on no-exec filesystems. */
diff -r -u -w mc-4.7.1-org/src/viewer/actions_cmd.c mc-4.7.1-viewertoeditor/src/viewer/actions_cmd.c
--- mc-4.7.1-org/src/viewer/actions_cmd.c	2010-02-26 15:01:54.000000000 +0300
+++ mc-4.7.1-viewertoeditor/src/viewer/actions_cmd.c	2010-04-26 02:19:54.000000000 +0400
@@ -439,6 +439,8 @@
         if (!mcview_is_in_panel (view))
             view->move_dir = (command == CK_ViewNextFile) ? 1 : -1;
         /* fallthrough */
+    case CK_ViewRunEditor:
+        view->run_editor = 1;
     case CK_ViewQuit:
         if (mcview_ok_to_quit (view))
             view->want_to_quit = TRUE;
diff -r -u -w mc-4.7.1-org/src/viewer/internal.h mc-4.7.1-viewertoeditor/src/viewer/internal.h
--- mc-4.7.1-org/src/viewer/internal.h	2010-02-26 15:01:54.000000000 +0300
+++ mc-4.7.1-viewertoeditor/src/viewer/internal.h	2010-04-26 02:02:33.000000000 +0400
@@ -191,6 +191,8 @@
     struct mcview_nroff_struct *search_nroff_seq;
 
     int search_numNeedSkipChar;
+
+    int run_editor;
 } mcview_t;
 
 typedef struct mcview_nroff_struct
diff -r -u -w mc-4.7.1-org/src/viewer/mcviewer.c mc-4.7.1-viewertoeditor/src/viewer/mcviewer.c
--- mc-4.7.1-org/src/viewer/mcviewer.c	2010-02-26 15:01:54.000000000 +0300
+++ mc-4.7.1-viewertoeditor/src/viewer/mcviewer.c	2010-04-26 02:04:35.000000000 +0400
@@ -257,6 +257,8 @@
     view->update_steps = 0;
     view->update_activate = 0;
 
+    view->run_editor = 0;
+
     if (mcview_default_hex_mode)
         mcview_toggle_hex_mode (view);
     if (mcview_default_nroff_flag)
@@ -275,7 +277,7 @@
 
 /* Real view only */
 int
-mcview_viewer (const char *command, const char *file, int *move_dir_p, int start_line)
+mcview_viewer (const char *command, const char *file, int *move_dir_p, int start_line, int * run_editor)
 {
     gboolean succeeded;
     mcview_t *lc_mcview;
@@ -296,6 +298,8 @@
         run_dlg (view_dlg);
         if (move_dir_p)
             *move_dir_p = lc_mcview->move_dir;
+        if (run_editor)
+            *run_editor = lc_mcview->run_editor;
     }
     else
     {
diff -r -u -w mc-4.7.1-org/src/viewer/mcviewer.h mc-4.7.1-viewertoeditor/src/viewer/mcviewer.h
--- mc-4.7.1-org/src/viewer/mcviewer.h	2010-02-26 15:01:54.000000000 +0300
+++ mc-4.7.1-viewertoeditor/src/viewer/mcviewer.h	2010-04-26 02:03:50.000000000 +0400
@@ -43,7 +43,7 @@
  * point to a variable that will receive the direction in which the user
  * wants to move (-1 = previous file, 1 = next file, 0 = do nothing).
  */
-extern int mcview_viewer (const char *command, const char *file, int *move_dir_p, int start_line);
+extern int mcview_viewer (const char *command, const char *file, int *move_dir_p, int start_line, int * run_editor);
 
 extern gboolean mcview_load (struct mcview_struct *, const char *, const char *, int);
 
