--- ./src/filemanager/file.c
+++ ./src/filemanager/file.c
@@ -2053,6 +2053,7 @@
         {
             dest_dir = d;
             d = NULL;
+            dest_dir_vpath = vfs_path_from_str (dest_dir);
             goto dont_mkdir;
         }
     }
@@ -2078,6 +2079,7 @@
     lp->dev = buf.st_dev;
     dest_dirs = g_slist_prepend (dest_dirs, lp);
 
+  dont_mkdir:
     if (ctx->preserve_uidgid)
     {
         while (mc_chown (dest_dir_vpath, cbuf.st_uid, cbuf.st_gid) != 0)
@@ -2096,7 +2098,6 @@
         }
     }
 
-  dont_mkdir:
     /* open the source dir for reading */
     reading = mc_opendir (src_vpath);
     if (reading == NULL)
