--- src/vfs/fish/helpers/ls.orig	2014-03-15 14:35:01.000000000 +0100
+++ src/vfs/fish/helpers/ls	2014-03-15 14:39:05.000000000 +0100
@@ -137,9 +137,13 @@
     if (S_ISLNK ($mode)) {
         my $linkname = readlink ("$dirname/$filename");
         $linkname =~ $strutils_shell_escape_regex;
-        printf("R%%o %%o $uid.$gid\nS$size\nd$mloctime\n:\"%%s\" -> \"%%s\"\n\n", S_IMODE($mode), S_IFMT($mode), $e_filename, $linkname);
+        printf("R%%o %%o %%i.%%i\nS%%i\nd%%s\n:\"%%s\" -> \"%%s\"\n\n",
+             S_IMODE($mode), S_IFMT($mode),
+             $uid, $gid, $size, $mloctime, $e_filename, $linkname);
     } else {
-        printf("R%%o %%o $uid.$gid\nS$size\nd$mloctime\n:\"%%s\"\n\n", S_IMODE($mode), S_IFMT($mode), $e_filename);
+        printf("R%%o %%o %%i.%%i\nS%%i\nd%%s\n:\"%%s\"\n\n",
+             S_IMODE($mode), S_IFMT($mode),
+             $uid, $gid, $size, $mloctime, $e_filename);
     }
 }
     printf("### 200\n");
