diff --git a/src/filemanager/find.c b/src/filemanager/find.c
index b28f8f7..fda5ac9 100644
--- a/src/filemanager/find.c
+++ b/src/filemanager/find.c
@@ -574,6 +574,20 @@ find_parameters (char **start_dir, char **pattern, char **content)
     widget_disable (content_regexp_cbox->widget, disable);
     add_widget (find_dlg, content_regexp_cbox);
 
+    content_use_cbox =
+        check_new (cbox_position--, FIND_X / 2 + 1, options.content_use, content_use_label);
+    add_widget (find_dlg, content_use_cbox);
+
+    in_with =
+        input_new (6, FIND_X / 2 + 1, input_get_default_colors (), FIND_X / 2 - 4, INPUT_LAST_TEXT,
+                   MC_HISTORY_SHARED_SEARCH, INPUT_COMPLETE_DEFAULT);
+    widget_disable (in_with->widget, disable);
+    add_widget (find_dlg, in_with);
+
+    content_label = label_new (5, FIND_X / 2 + 1, _("Content:"));
+    widget_disable (content_label->widget, disable);
+    add_widget (find_dlg, content_label);
+
     cbox_position = FIND_Y - 6;
 
     skip_hidden_cbox = check_new (cbox_position--, 3, options.skip_hidden, file_skip_hidden_label);
@@ -594,21 +608,6 @@ find_parameters (char **start_dir, char **pattern, char **content)
     recursively_cbox = check_new (cbox_position, 3, options.find_recurs, file_recurs_label);
     add_widget (find_dlg, recursively_cbox);
 
-    /* This checkbox is located in the second column */
-    content_use_cbox =
-        check_new (cbox_position, FIND_X / 2 + 1, options.content_use, content_use_label);
-    add_widget (find_dlg, content_use_cbox);
-
-    in_with =
-        input_new (6, FIND_X / 2 + 1, input_get_default_colors (), FIND_X / 2 - 4, INPUT_LAST_TEXT,
-                   MC_HISTORY_SHARED_SEARCH, INPUT_COMPLETE_DEFAULT);
-    widget_disable (in_with->widget, disable);
-    add_widget (find_dlg, in_with);
-
-    content_label = label_new (5, FIND_X / 2 + 1, _("Content:"));
-    widget_disable (content_label->widget, disable);
-    add_widget (find_dlg, content_label);
-
     in_name = input_new (6, 3, input_get_default_colors (),
                          FIND_X / 2 - 4, INPUT_LAST_TEXT, "name", INPUT_COMPLETE_DEFAULT);
     add_widget (find_dlg, in_name);
