diff --git a/lib/widget/input.c b/lib/widget/input.c
index 4ce2ddb..c8d7b74 100644
--- a/lib/widget/input.c
+++ b/lib/widget/input.c
@@ -425,7 +425,7 @@ backward_word (WInput * in)
     const char *p, *p_tmp;
 
     for (p = in->buffer + str_offset_to_pos (in->buffer, in->point);
-         (p != in->buffer) && (p[0] == '\0'); str_cprev_char (&p), in->point--);
+         (p != in->buffer) && (p[-1] == '\0'); str_cprev_char (&p), in->point--);
 
     while (p != in->buffer)
     {
