From b40f8102717c89e72ee76e3c36adab4540df5447 Mon Sep 17 00:00:00 2001
From: Whisperity <whisperity@gmail.com>
Date: Sat, 30 Jul 2022 11:54:43 +0200
Subject: [PATCH] (tty_check_term): support Contour.

---
 lib/tty/tty.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/lib/tty/tty.c b/lib/tty/tty.c
index f27af9cca..60b138e91 100644
--- a/lib/tty/tty.c
+++ b/lib/tty/tty.c
@@ -101,6 +101,7 @@ sigintr_handler (int signo)
  *         foot*
  *         screen*
  *         tmux*
+ *         contour*
  */
 gboolean
 tty_check_term (gboolean force_xterm)
@@ -124,7 +125,8 @@ tty_check_term (gboolean force_xterm)
         || strncmp (termvalue, "alacritty", 9) == 0
         || strncmp (termvalue, "foot", 4) == 0
         || strncmp (termvalue, "screen", 6) == 0
-        || strncmp (termvalue, "tmux", 4) == 0;
+        || strncmp (termvalue, "tmux", 4) == 0
+        || strncmp (termvalue, "contour", 7) == 0;
     /* *INDENT-ON* */
 }
 
-- 
2.25.1

