--- util.c.orig	2023-08-06 15:26:05.000000000 +0200
+++ util.c	2024-01-25 18:33:57.745736014 +0100
@@ -453,12 +453,13 @@
 #endif
     };
     /* *INDENT-ON* */
-    static const char *const suffix[] =
-        { "", "K", "M", "G", "T", "P", "E", "Z", "Y", "R", "Q", NULL };
-    static const char *const suffix_lc[] =
-        { "", "k", "m", "g", "t", "p", "e", "z", "y", "r", "q", NULL };
+    static const char *const suffix_iec[] =
+        { "", "Ki", "Mi", "Gi", "Ti", "Pi", "Ei", "Zi", "Yi", "Ri",
+          "Qi", NULL };
+    static const char *const suffix_si[] =
+        { "", "k", "M", "G", "T", "P", "E", "Z", "Y", "R", "Q", NULL };
 
-    const char *const *sfx = use_si ? suffix_lc : suffix;
+    const char *const *sfx = use_si ? suffix_si : suffix_iec;
     int j = 0;
 
     if (len == 0)
