--- /home/users/gotar/git/mc/src/viewer/actions_cmd.c	2010-11-20 13:22:07.000000000 +0100
+++ actions_cmd.c	2010-11-21 00:55:51.000000000 +0100
@@ -242,7 +242,7 @@ mcview_handle_editkey (mcview_t * view, 
 
 /* --------------------------------------------------------------------------------------------- */
 
-static cb_ret_t
+static int
 mcview_execute_cmd (mcview_t * view, unsigned long command)
 {
     int res = MSG_HANDLED;
@@ -376,7 +376,7 @@ mcview_execute_cmd (mcview_t * view, uns
     case CK_ViewQuit:
         if (!mcview_is_in_panel (view))
             dlg_stop (view->widget.owner);
-        break;
+        return -1;
     default:
         res = MSG_NOT_HANDLED;
     }
@@ -384,10 +384,11 @@ mcview_execute_cmd (mcview_t * view, uns
 }
 
 /* Both views */
-static cb_ret_t
+static int
 mcview_handle_key (mcview_t * view, int key)
 {
     unsigned long command;
+    int res;
 
     key = convert_from_input_c (key);
 
@@ -402,8 +403,12 @@ mcview_handle_key (mcview_t * view, int 
     }
 
     command = lookup_keymap_command (view->plain_map, key);
-    if ((command != CK_Ignore_Key) && (mcview_execute_cmd (view, command) == MSG_HANDLED))
-        return MSG_HANDLED;
+    if (command != CK_Ignore_Key)
+    {
+        res = mcview_execute_cmd (view, command);
+        if (res != MSG_NOT_HANDLED)
+            return res;
+    }
 
 #ifdef MC_ENABLE_DEBUGGING_CODE
     if (c == 't')
@@ -450,7 +455,7 @@ cb_ret_t
 mcview_callback (Widget * w, widget_msg_t msg, int parm)
 {
     mcview_t *view = (mcview_t *) w;
-    cb_ret_t i;
+    int i;
 
     mcview_compute_areas (view);
     mcview_update_bytes_per_line (view);
@@ -475,7 +480,10 @@ mcview_callback (Widget * w, widget_msg_
 
     case WIDGET_KEY:
         i = mcview_handle_key (view, parm);
-        mcview_update (view);
+        if (i == MSG_HANDLED)
+            mcview_update (view);
+        if (i == -1)
+            i = MSG_HANDLED;
         return i;
 
     case WIDGET_COMMAND:
