diff -urN mc-4.7.1-old/lib/global.h mc-4.7.1-new/lib/global.h
--- mc-4.7.1-old/lib/global.h	2010-02-25 10:51:39.000000000 +0000
+++ mc-4.7.1-new/lib/global.h	2010-03-17 20:17:40.000000000 +0000
@@ -114,8 +114,6 @@
 #define BUF_SMALL	128
 #define BUF_TINY	64
 
-void refresh_screen (void *);
-
 /* AIX compiler doesn't understand '\e' */
 #define ESC_CHAR '\033'
 #define ESC_STR  "\033"
diff -urN mc-4.7.1-old/lib/mcconfig/get.c mc-4.7.1-new/lib/mcconfig/get.c
--- mc-4.7.1-old/lib/mcconfig/get.c	2010-02-25 10:51:39.000000000 +0000
+++ mc-4.7.1-new/lib/mcconfig/get.c	2010-03-17 19:23:40.000000000 +0000
@@ -202,29 +202,3 @@
 }
 
 /* - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - */
-
-gboolean *
-mc_config_get_bool_list (mc_config_t * mc_config, const gchar * group,
-			 const gchar * param, gsize * length)
-{
-    if (!mc_config || !group || !param)
-	return NULL;
-
-    return g_key_file_get_boolean_list (mc_config->handle, group, param,
-					length, NULL);
-}
-
-/* - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - */
-
-int *
-mc_config_get_int_list (mc_config_t * mc_config, const gchar * group,
-			const gchar * param, gsize * length)
-{
-    if (!mc_config || !group || !param)
-	return NULL;
-
-    return g_key_file_get_integer_list (mc_config->handle, group, param,
-					length, NULL);
-}
-
-/* - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - */
diff -urN mc-4.7.1-old/lib/mcconfig/set.c mc-4.7.1-new/lib/mcconfig/set.c
--- mc-4.7.1-old/lib/mcconfig/set.c	2010-02-25 10:51:39.000000000 +0000
+++ mc-4.7.1-new/lib/mcconfig/set.c	2010-03-17 19:25:02.000000000 +0000
@@ -144,30 +144,3 @@
 }
 
 /* - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - */
-
-
-void
-mc_config_set_bool_list (mc_config_t * mc_config, const gchar * group,
-			 const gchar * param, gboolean value[], gsize length)
-{
-    if (!mc_config || !group || !param || !value || length == 0)
-	return;
-
-    g_key_file_set_boolean_list (mc_config->handle, group, param, value,
-				 length);
-}
-
-/* - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - */
-
-void
-mc_config_set_int_list (mc_config_t * mc_config, const gchar * group,
-			const gchar * param, int value[], gsize length)
-{
-    if (!mc_config || !group || !param || !value || length == 0)
-	return;
-
-    g_key_file_set_integer_list (mc_config->handle, group, param, value,
-				 length);
-}
-
-/* - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - */
diff -urN mc-4.7.1-old/lib/mcconfig.h mc-4.7.1-new/lib/mcconfig.h
--- mc-4.7.1-old/lib/mcconfig.h	2010-02-25 10:51:39.000000000 +0000
+++ mc-4.7.1-new/lib/mcconfig.h	2010-03-17 19:25:13.000000000 +0000
@@ -57,10 +57,6 @@
 
 gchar **mc_config_get_string_list (mc_config_t *, const gchar *, const gchar *, gsize *);
 
-gboolean *mc_config_get_bool_list (mc_config_t *, const gchar *, const gchar *, gsize *);
-
-int *mc_config_get_int_list (mc_config_t *, const gchar *, const gchar *, gsize *);
-
 
 /* mcconfig/set.c: */
 
@@ -79,12 +75,6 @@
   mc_config_set_string_list (mc_config_t *, const gchar *,
                              const gchar *, const gchar * const[], gsize);
 
-void
-  mc_config_set_bool_list (mc_config_t *, const gchar *, const gchar *, gboolean[], gsize);
-
-void
-  mc_config_set_int_list (mc_config_t *, const gchar *, const gchar *, int[], gsize);
-
 
 /* mcconfig/dialog.c: */
 
diff -urN mc-4.7.1-old/lib/search/search.c mc-4.7.1-new/lib/search/search.c
--- mc-4.7.1-old/lib/search/search.c	2010-02-25 10:51:39.000000000 +0000
+++ mc-4.7.1-new/lib/search/search.c	2010-03-17 19:27:35.000000000 +0000
@@ -387,47 +387,3 @@
 }
 
 /* --------------------------------------------------------------------------------------------- */
-
-int
-mc_search_getstart_rezult_by_num (mc_search_t * lc_mc_search, int lc_index)
-{
-    if (!lc_mc_search)
-        return 0;
-    if (lc_mc_search->search_type == MC_SEARCH_T_NORMAL)
-        return 0;
-#ifdef SEARCH_TYPE_GLIB
-    {
-        gint start_pos;
-        gint end_pos;
-        g_match_info_fetch_pos (lc_mc_search->regex_match_info, lc_index, &start_pos, &end_pos);
-        return (int) start_pos;
-    }
-#else /* SEARCH_TYPE_GLIB */
-    return lc_mc_search->iovector[lc_index * 2];
-#endif /* SEARCH_TYPE_GLIB */
-
-}
-
-/* --------------------------------------------------------------------------------------------- */
-
-int
-mc_search_getend_rezult_by_num (mc_search_t * lc_mc_search, int lc_index)
-{
-    if (!lc_mc_search)
-        return 0;
-    if (lc_mc_search->search_type == MC_SEARCH_T_NORMAL)
-        return 0;
-#ifdef SEARCH_TYPE_GLIB
-    {
-        gint start_pos;
-        gint end_pos;
-        g_match_info_fetch_pos (lc_mc_search->regex_match_info, lc_index, &start_pos, &end_pos);
-        return (int) end_pos;
-    }
-#else /* SEARCH_TYPE_GLIB */
-    return lc_mc_search->iovector[lc_index * 2 + 1];
-#endif /* SEARCH_TYPE_GLIB */
-
-}
-
-/* --------------------------------------------------------------------------------------------- */
diff -urN mc-4.7.1-old/lib/search.h mc-4.7.1-new/lib/search.h
--- mc-4.7.1-old/lib/search.h	2010-02-25 10:51:39.000000000 +0000
+++ mc-4.7.1-new/lib/search.h	2010-03-17 19:27:41.000000000 +0000
@@ -140,7 +140,4 @@
 
 gboolean mc_search (const gchar *, const gchar *, mc_search_type_t);
 
-int mc_search_getstart_rezult_by_num (mc_search_t *, int);
-int mc_search_getend_rezult_by_num (mc_search_t *, int);
-
 #endif
diff -urN mc-4.7.1-old/lib/skin/internal.h mc-4.7.1-new/lib/skin/internal.h
--- mc-4.7.1-old/lib/skin/internal.h	2010-02-25 10:51:39.000000000 +0000
+++ mc-4.7.1-new/lib/skin/internal.h	2010-03-17 20:17:20.000000000 +0000
@@ -26,7 +26,6 @@
 gboolean mc_skin_ini_file_parse (mc_skin_t *);
 void mc_skin_set_hardcoded_skin (mc_skin_t *);
 
-gboolean mc_skin_ini_file_parse_colors (mc_skin_t *);
 gboolean mc_skin_color_parse_ini_file (mc_skin_t *);
 
 void mc_skin_hardcoded_ugly_lines (mc_skin_t *);
diff -urN mc-4.7.1-old/lib/strescape.h mc-4.7.1-new/lib/strescape.h
--- mc-4.7.1-old/lib/strescape.h	2010-02-25 10:51:39.000000000 +0000
+++ mc-4.7.1-new/lib/strescape.h	2010-03-17 19:44:29.000000000 +0000
@@ -22,10 +22,8 @@
 char *strutils_shell_escape (const char *);
 
 char *strutils_glob_escape (const char *);
-char *strutils_glob_unescape (const char *);
 
 char *strutils_regex_escape (const char *);
-char *strutils_regex_unescape (const char *);
 
 gboolean strutils_is_char_escaped (const char *, const char *);
 
diff -urN mc-4.7.1-old/lib/strutil/strescape.c mc-4.7.1-new/lib/strutil/strescape.c
--- mc-4.7.1-old/lib/strutil/strescape.c	2010-02-25 10:51:39.000000000 +0000
+++ mc-4.7.1-new/lib/strutil/strescape.c	2010-03-17 19:44:22.000000000 +0000
@@ -201,21 +201,6 @@
 
 /* --------------------------------------------------------------------------------------------- */
 
-char *
-strutils_glob_unescape (const char *text)
-{
-    return strutils_unescape (text, -1, ESCAPE_GLOB_CHARS, TRUE);
-}
-
-/* --------------------------------------------------------------------------------------------- */
-char *
-strutils_regex_unescape (const char *text)
-{
-    return strutils_unescape (text, -1, ESCAPE_REGEX_CHARS, TRUE);
-}
-
-/* --------------------------------------------------------------------------------------------- */
-
 /** Check if char in pointer contain escape'd chars
 
  \param start
diff -urN mc-4.7.1-old/lib/strutil/strutil.c mc-4.7.1-new/lib/strutil/strutil.c
--- mc-4.7.1-old/lib/strutil/strutil.c	2010-02-25 10:51:39.000000000 +0000
+++ mc-4.7.1-new/lib/strutil/strutil.c	2010-03-17 19:43:15.000000000 +0000
@@ -293,12 +293,6 @@
     va_end (ap);
 }
 
-void
-str_insert_replace_char (GString * buffer)
-{
-    used_class.insert_replace_char (buffer);
-}
-
 estr_t
 str_translate_char (GIConv conv, const char *keys, size_t ch_size,
 		    char *output, size_t out_size)
@@ -492,13 +486,6 @@
     used_class.cprev_char (text);
 }
 
-char *
-str_get_next_char_safe (char *text)
-{
-    used_class.cnext_char_safe ((const char **) &text);
-    return text;
-}
-
 const char *
 str_cget_next_char_safe (const char *text)
 {
@@ -506,50 +493,6 @@
     return text;
 }
 
-void
-str_next_char_safe (char **text)
-{
-    used_class.cnext_char_safe ((const char **) text);
-}
-
-void
-str_cnext_char_safe (const char **text)
-{
-    used_class.cnext_char_safe (text);
-}
-
-char *
-str_get_prev_char_safe (char *text)
-{
-    used_class.cprev_char_safe ((const char **) &text);
-    return text;
-}
-
-const char *
-str_cget_prev_char_safe (const char *text)
-{
-    used_class.cprev_char_safe (&text);
-    return text;
-}
-
-void
-str_prev_char_safe (char **text)
-{
-    used_class.cprev_char_safe ((const char **) text);
-}
-
-void
-str_cprev_char_safe (const char **text)
-{
-    used_class.cprev_char_safe (text);
-}
-
-int
-str_next_noncomb_char (char **text)
-{
-    return used_class.cnext_noncomb_char ((const char **) text);
-}
-
 int
 str_cnext_noncomb_char (const char **text)
 {
@@ -587,12 +530,6 @@
 }
 
 int
-str_term_char_width (const char *text)
-{
-    return used_class.term_char_width (text);
-}
-
-int
 str_offset_to_pos (const char *text, size_t length)
 {
     return used_class.offset_to_pos (text, length);
@@ -611,18 +548,6 @@
 }
 
 int
-str_length2 (const char *text, int size)
-{
-    return used_class.length2 (text, size);
-}
-
-int
-str_length_noncomb (const char *text)
-{
-    return used_class.length_noncomb (text);
-}
-
-int
 str_column_to_pos (const char *text, size_t pos)
 {
     return used_class.column_to_pos (text, pos);
@@ -670,12 +595,6 @@
     return used_class.isprint (ch);
 }
 
-int
-str_iscombiningmark (const char *ch)
-{
-    return used_class.iscombiningmark (ch);
-}
-
 const char *
 str_trunc (const char *text, int width)
 {
@@ -701,12 +620,6 @@
     return used_class.search_first (text, search, case_sen);
 }
 
-const char *
-str_search_last (const char *text, const char *search, int case_sen)
-{
-    return used_class.search_last (text, search, case_sen);
-}
-
 int
 str_is_valid_string (const char *text)
 {
@@ -714,47 +627,11 @@
 }
 
 int
-str_compare (const char *t1, const char *t2)
-{
-    return used_class.compare (t1, t2);
-}
-
-int
-str_ncompare (const char *t1, const char *t2)
-{
-    return used_class.ncompare (t1, t2);
-}
-
-int
 str_casecmp (const char *t1, const char *t2)
 {
     return used_class.casecmp (t1, t2);
 }
 
-int
-str_ncasecmp (const char *t1, const char *t2)
-{
-    return used_class.ncasecmp (t1, t2);
-}
-
-int
-str_prefix (const char *text, const char *prefix)
-{
-    return used_class.prefix (text, prefix);
-}
-
-int
-str_caseprefix (const char *text, const char *prefix)
-{
-    return used_class.caseprefix (text, prefix);
-}
-
-void
-str_fix_string (char *text)
-{
-    used_class.fix_string (text);
-}
-
 char *
 str_create_key (const char *text, int case_sen)
 {
diff -urN mc-4.7.1-old/lib/strutil.h mc-4.7.1-new/lib/strutil.h
--- mc-4.7.1-old/lib/strutil.h	2010-02-26 11:34:02.000000000 +0000
+++ mc-4.7.1-new/lib/strutil.h	2010-03-17 19:43:25.000000000 +0000
@@ -188,10 +188,6 @@
 void
 str_printf (GString *, const char *, ...);
 
-/* add standard replacement character in terminal encoding
- */
-void str_insert_replace_char (GString *);
-
 /* init strings and set terminal encoding,
  * if is termenc NULL, detect terminal encoding
  * create all str_cnv_* and set functions for terminal encoding
@@ -249,36 +245,8 @@
  * works with invalid string 
  * I
  */
-char *str_get_next_char_safe (char *text);
 const char *str_cget_next_char_safe (const char *text);
 
-/* return previous characters before text, do not call on the start of strings
- * works with invalid string 
- * I
- */
-char *str_get_prev_char_safe (char *text);
-const char *str_cget_prev_char_safe (const char *text);
-
-/* set text to next characters, do not call on the end of string
- * works with invalid string 
- * I
- */
-void str_next_char_safe (char **text);
-void str_cnext_char_safe (const char **text);
-
-/* set text to previous characters, do not call on the start of strings
- * works with invalid string 
- * I
- */
-void str_prev_char_safe (char **text);
-void str_cprev_char_safe (const char **text);
-
-/* set text to next noncombining characters, check the end of text
- * return how many characters was skipped
- * works with invalid string 
- * I
- */
-int str_next_noncomb_char (char **text);
 int str_cnext_noncomb_char (const char **text);
 
 /* set text to previous noncombining characters, search stop at begin 
@@ -314,12 +282,6 @@
  */
 int str_isprint (const char *ch);
 
-/* if first characters in ch is a combining mark (only in utf-8)
- * combining makrs are assumed to be zero width 
- * I
- */
-int str_iscombiningmark (const char *ch);
-
 /* write lower from of fisrt characters in ch into out
  * decrase remain by size of returned characters
  * if out is not big enough, do nothing
@@ -337,27 +299,11 @@
  */
 int str_length (const char* text);
 
-/* return length of text in characters, limit to size
- * I
- */
-int str_length2 (const char* text, int size);
-
 /* return length of one char
  * I
  */
 int str_length_char (const char *);
 
-/* return length of text in characters, count only noncombining characters
- * I
- */
-int str_length_noncomb (const char* text);
-
-/* replace all invalid characters in text with questionmark
- * after return, text is valid string in terminal encoding
- * I
- */
-void str_fix_string (char* text);
-
 /* replace all invalid characters in text with questionmark
  * replace all unprintable characters with '.'
  * return static allocated string, "text" is not changed
@@ -403,12 +349,6 @@
  */
 int str_term_width2 (const char *text, size_t length);
 
-/* return width, that will be character occupied on terminal
- * combining characters are always zero width
- * I
- */
-int str_term_char_width (const char *text);
-
 /* convert position in characters to position in bytes 
  * I
  */
@@ -439,45 +379,11 @@
  */
 const char *str_search_first (const char *text, const char *needle, int case_sen);
 
-/* search for last occurrence of search in text
- */
-const char *str_search_last (const char *text, const char *needle, int case_sen);
-
-/* case sensitive compare two strings
- * I
- */
-int str_compare (const char *t1, const char *t2);
-
-/* case sensitive compare two strings
- * if one string is prefix of the other string, return 0
- * I
- */
-int str_ncompare (const char *t1, const char *t2);
-
 /* case insensitive compare two strings
  * I
  */
 int str_casecmp (const char *t1, const char *t2);
 
-/* case insensitive compare two strings
- * if one string is prefix of the other string, return 0
- * I
- */
-int str_ncasecmp (const char *t1, const char *t2);
-
-/* return, how many bytes are are same from start in text and prefix
- * both strings are decomposed befor comapring and return value is counted
- * in decomposed form, too. caling with prefix, prefix, you get size in bytes
- * of prefix in decomposed form,
- * I
- */
-int str_prefix (const char *text, const char *prefix);
-
-/* case insensitive version of str_prefix
- * I
- */
-int str_caseprefix (const char *text, const char *prefix);
-
 /* create a key that is used by str_key_collate
  * I
  */
diff -urN mc-4.7.1-old/lib/tty/color.c mc-4.7.1-new/lib/tty/color.c
--- mc-4.7.1-old/lib/tty/color.c	2010-02-26 12:22:31.000000000 +0000
+++ mc-4.7.1-new/lib/tty/color.c	2010-03-17 19:44:59.000000000 +0000
@@ -239,14 +239,6 @@
 /* --------------------------------------------------------------------------------------------- */
 
 void
-tty_color_free_all_non_tmp (void)
-{
-    tty_color_free_all (FALSE);
-}
-
-/* --------------------------------------------------------------------------------------------- */
-
-void
 tty_color_set_defaults (const char *fgcolor, const char *bgcolor)
 {
     g_free (tty_color_defaults__fg);
diff -urN mc-4.7.1-old/lib/tty/color.h mc-4.7.1-new/lib/tty/color.h
--- mc-4.7.1-old/lib/tty/color.h	2010-02-26 09:29:51.000000000 +0000
+++ mc-4.7.1-new/lib/tty/color.h	2010-03-17 19:44:49.000000000 +0000
@@ -31,7 +31,6 @@
 int tty_try_alloc_color_pair2 (const char *, const char *, gboolean);
 
 void tty_color_free_all_tmp (void);
-void tty_color_free_all_non_tmp (void);
 
 void tty_setcolor (int color);
 void tty_lowlevel_setcolor (int color);
diff -urN mc-4.7.1-old/lib/tty/key.h mc-4.7.1-new/lib/tty/key.h
--- mc-4.7.1-old/lib/tty/key.h	2010-02-25 10:51:39.000000000 +0000
+++ mc-4.7.1-new/lib/tty/key.h	2010-03-17 20:17:57.000000000 +0000
@@ -63,7 +63,6 @@
 /* Channel manipulation */
 void add_select_channel (int fd, select_fn callback, void *info);
 void delete_select_channel (int fd);
-void remove_select_channel (int fd);
 
 /* Activate/deactivate the channel checking */
 void channels_up (void);
diff -urN mc-4.7.1-old/lib/tty/tty.c mc-4.7.1-new/lib/tty/tty.c
--- mc-4.7.1-old/lib/tty/tty.c	2010-02-26 09:29:51.000000000 +0000
+++ mc-4.7.1-new/lib/tty/tty.c	2010-03-17 19:45:25.000000000 +0000
@@ -126,12 +126,6 @@
 }
 
 void
-tty_print_one_hline (void)
-{
-    tty_print_alt_char (mc_tty_ugly_frm[MC_TTY_FRM_thinhoriz]);
-}
-
-void
 tty_print_one_vline (void)
 {
     tty_print_alt_char (mc_tty_ugly_frm[MC_TTY_FRM_thinvert]);
diff -urN mc-4.7.1-old/lib/tty/tty.h mc-4.7.1-new/lib/tty/tty.h
--- mc-4.7.1-old/lib/tty/tty.h	2010-02-26 09:29:51.000000000 +0000
+++ mc-4.7.1-new/lib/tty/tty.h	2010-03-17 19:45:18.000000000 +0000
@@ -72,7 +72,6 @@
 extern void tty_printf (const char *s, ...);
 
 extern void tty_print_one_vline (void);
-extern void tty_print_one_hline (void);
 extern void tty_draw_hline (int y, int x, int ch, int len);
 extern void tty_draw_vline (int y, int x, int ch, int len);
 extern void tty_draw_box (int y, int x, int rows, int cols);
diff -urN mc-4.7.1-old/lib/util.c mc-4.7.1-new/lib/util.c
--- mc-4.7.1-old/lib/util.c	2010-02-26 12:01:53.000000000 +0000
+++ mc-4.7.1-new/lib/util.c	2010-03-17 19:17:31.000000000 +0000
@@ -1588,20 +1588,6 @@
 }
 
 #undef TMP_SUFFIX
-extern const char *
-cstrcasestr (const char *haystack, const char *needle)
-{
-    char *nee = str_create_search_needle (needle, 0);
-    const char *result = str_search_first (haystack, nee, 0);
-    str_release_search_needle (nee, 0);
-    return result;
-}
-
-const char *
-cstrstr (const char *haystack, const char *needle)
-{
-    return strstr (haystack, needle);
-}
 
 extern char *
 str_unconst (const char *s)
diff -urN mc-4.7.1-old/lib/util.h mc-4.7.1-new/lib/util.h
--- mc-4.7.1-old/lib/util.h	2010-02-26 11:34:02.000000000 +0000
+++ mc-4.7.1-new/lib/util.h	2010-03-17 19:17:22.000000000 +0000
@@ -22,9 +22,6 @@
 
 /* String managing functions */
 
-extern const char *cstrcasestr (const char *haystack, const char *needle);
-extern const char *cstrstr (const char *haystack, const char *needle);
-
 void str_replace(char *s, char from, char to);
 int  is_printable (int c);
 void msglen (const char *text, /*@out@*/ int *lines, /*@out@*/ int *columns);
diff -urN mc-4.7.1-old/lib/vfs/mc-vfs/gc.c mc-4.7.1-new/lib/vfs/mc-vfs/gc.c
--- mc-4.7.1-old/lib/vfs/mc-vfs/gc.c	2010-02-25 10:51:39.000000000 +0000
+++ mc-4.7.1-new/lib/vfs/mc-vfs/gc.c	2010-03-17 18:39:19.000000000 +0000
@@ -271,7 +271,7 @@
  * seconds until the next item times out.
  */
 int
-vfs_timeouts ()
+vfs_timeouts (void)
 {
     return stamps ? 10 : 0;
 }
diff -urN mc-4.7.1-old/lib/vfs/mc-vfs/vfs-impl.h mc-4.7.1-new/lib/vfs/mc-vfs/vfs-impl.h
--- mc-4.7.1-old/lib/vfs/mc-vfs/vfs-impl.h	2010-02-25 10:51:39.000000000 +0000
+++ mc-4.7.1-new/lib/vfs/mc-vfs/vfs-impl.h	2010-03-17 20:18:12.000000000 +0000
@@ -127,7 +127,6 @@
 
 struct vfs_class *vfs_get_class (const char *path);
 struct vfs_class *vfs_split (char *path, char **inpath, char **op);
-char *vfs_path (const char *path);
 int vfs_file_class_flags (const char *filename);
 
 void vfs_fill_names (fill_names_f);
diff -urN mc-4.7.1-old/src/charsets.h mc-4.7.1-new/src/charsets.h
--- mc-4.7.1-old/src/charsets.h	2010-02-25 10:51:40.000000000 +0000
+++ mc-4.7.1-new/src/charsets.h	2010-03-17 20:15:07.000000000 +0000
@@ -31,7 +31,6 @@
 const char *init_translation_table (int cpsource, int cpdisplay);
 void convert_to_display (char *str);
 void convert_from_input (char *str);
-void convert_string (unsigned char *str);
 /*
  * Converter from utf to selected codepage
  * param str, utf char
diff -urN mc-4.7.1-old/src/ecs.c mc-4.7.1-new/src/ecs.c
--- mc-4.7.1-old/src/ecs.c	2010-02-25 10:51:40.000000000 +0000
+++ mc-4.7.1-new/src/ecs.c	2010-03-17 19:20:13.000000000 +0000
@@ -104,61 +104,11 @@
 	(cf(c))
 #endif
 
-extern gboolean ecs_isalnum(ecs_char c)
-{
-	return ECS_CTYPE(iswalnum, isalnum, c);
-}
-
-extern gboolean ecs_isalpha(ecs_char c)
-{
-	return ECS_CTYPE(iswalpha, isalpha, c);
-}
-
-extern gboolean ecs_iscntrl(ecs_char c)
-{
-	return ECS_CTYPE(iswcntrl, iscntrl, c);
-}
-
-extern gboolean ecs_isdigit(ecs_char c)
-{
-	return ECS_CTYPE(iswdigit, isdigit, c);
-}
-
-extern gboolean ecs_isgraph(ecs_char c)
-{
-	return ECS_CTYPE(iswgraph, isgraph, c);
-}
-
-extern gboolean ecs_islower(ecs_char c)
-{
-	return ECS_CTYPE(iswlower, islower, c);
-}
-
 extern gboolean ecs_isprint(ecs_char c)
 {
 	return ECS_CTYPE(iswprint, isprint, c);
 }
 
-extern gboolean ecs_ispunct(ecs_char c)
-{
-	return ECS_CTYPE(iswpunct, ispunct, c);
-}
-
-extern gboolean ecs_isspace(ecs_char c)
-{
-	return ECS_CTYPE(iswspace, isspace, c);
-}
-
-extern gboolean ecs_isupper(ecs_char c)
-{
-	return ECS_CTYPE(iswupper, isupper, c);
-}
-
-extern gboolean ecs_isxdigit(ecs_char c)
-{
-	return ECS_CTYPE(iswxdigit, isxdigit, c);
-}
-
 #undef ECS_CTYPE
 
 /*
@@ -231,18 +181,6 @@
 	}
 }
 
-size_t
-ecs_strspn(const ecs_char *s, const ecs_char *chars)
-{
-	size_t i;
-
-	for (i = 0; s[i] != ECS_CHAR('\0'); i++) {
-		if (ecs_strchr(chars, s[i]) == NULL)
-			break;
-	}
-	return i;
-}
-
 ecs_char *
 ecs_strstr(const ecs_char *s, const ecs_char *sub)
 {
diff -urN mc-4.7.1-old/src/ecs.h mc-4.7.1-new/src/ecs.h
--- mc-4.7.1-old/src/ecs.h	2010-02-25 10:51:40.000000000 +0000
+++ mc-4.7.1-new/src/ecs.h	2010-03-17 19:19:59.000000000 +0000
@@ -66,17 +66,7 @@
  * Replacements for the ISO C90 <ctype.h> functions.
  */
 
-extern gboolean ecs_isalnum(ecs_char);
-extern gboolean ecs_isalpha(ecs_char);
-extern gboolean ecs_iscntrl(ecs_char);
-extern gboolean ecs_isdigit(ecs_char);
-extern gboolean ecs_isgraph(ecs_char);
-extern gboolean ecs_islower(ecs_char);
 extern gboolean ecs_isprint(ecs_char);
-extern gboolean ecs_ispunct(ecs_char);
-extern gboolean ecs_isspace(ecs_char);
-extern gboolean ecs_isupper(ecs_char);
-extern gboolean ecs_isxdigit(ecs_char);
 
 /*
  * Replacements for the ISO C90 <string.h> functions.
@@ -94,7 +84,6 @@
 extern size_t ecs_strcspn(const ecs_char *, const ecs_char *);
 /* left out: ecs_strpbrk */
 extern ecs_char *ecs_strrchr(const ecs_char *, ecs_char);
-extern size_t ecs_strspn(const ecs_char *, const ecs_char *);
 extern ecs_char *ecs_strstr(const ecs_char *, const ecs_char *);
 /* left out: ecs_strtok */
 extern size_t ecs_strlen(const ecs_char *);
diff -urN mc-4.7.1-old/src/editor/edit-impl.h mc-4.7.1-new/src/editor/edit-impl.h
--- mc-4.7.1-old/src/editor/edit-impl.h	2010-02-25 10:51:40.000000000 +0000
+++ mc-4.7.1-new/src/editor/edit-impl.h	2010-03-17 20:15:39.000000000 +0000
@@ -235,7 +235,6 @@
 int edit_load_forward_cmd (WEdit * edit);
 void edit_block_process_cmd (WEdit * edit, const char *shell_cmd, int block);
 void edit_refresh_cmd (WEdit * edit);
-void edit_date_cmd (WEdit * edit);
 void edit_goto_cmd (WEdit * edit);
 int eval_marks (WEdit * edit, long *start_mark, long *end_mark);
 void edit_status (WEdit * edit);
@@ -245,7 +244,6 @@
 int edit_print_string (WEdit * e, const char *s);
 void edit_move_to_line (WEdit * e, long line);
 void edit_move_display (WEdit * e, long line);
-void edit_word_wrap (WEdit * edit);
 int edit_sort_cmd (WEdit * edit);
 int edit_ext_cmd (WEdit * edit);
 void edit_help_cmd (WEdit * edit);
@@ -274,7 +272,6 @@
 
 void book_mark_insert (WEdit * edit, int line, int c);
 int book_mark_query_color (WEdit * edit, int line, int c);
-int book_mark_query_all (WEdit * edit, int line, int *c);
 struct _book_mark *book_mark_find (WEdit * edit, int line);
 int book_mark_clear (WEdit * edit, int line, int c);
 void book_mark_flush (WEdit * edit, int c);
diff -urN mc-4.7.1-old/src/setup.c mc-4.7.1-new/src/setup.c
--- mc-4.7.1-old/src/setup.c	2010-02-26 12:01:54.000000000 +0000
+++ mc-4.7.1-new/src/setup.c	2010-03-17 18:22:19.000000000 +0000
@@ -474,7 +474,7 @@
 }
 
 static void
-load_layout ()
+load_layout (void)
 {
     int i;
 
diff -urN mc-4.7.1-old/src/viewer/internal.h mc-4.7.1-new/src/viewer/internal.h
--- mc-4.7.1-old/src/viewer/internal.h	2010-02-26 12:01:54.000000000 +0000
+++ mc-4.7.1-new/src/viewer/internal.h	2010-03-17 20:19:01.000000000 +0000
@@ -312,7 +312,7 @@
 
 /* nroff.c: */
 void mcview_display_nroff (mcview_t * view);
-int mcview__get_nroff_real_len (mcview_t * view, off_t, off_t p);
+int mcview_get_nroff_real_len (mcview_t * view, off_t, off_t p);
 
 mcview_nroff_t *mcview_nroff_seq_new_num (mcview_t * view, off_t p);
 mcview_nroff_t *mcview_nroff_seq_new (mcview_t * view);
diff -urN mc-4.7.1-old/src/viewer/nroff.c mc-4.7.1-new/src/viewer/nroff.c
--- mc-4.7.1-old/src/viewer/nroff.c	2010-02-26 12:01:54.000000000 +0000
+++ mc-4.7.1-new/src/viewer/nroff.c	2010-03-17 20:18:51.000000000 +0000
@@ -220,7 +220,7 @@
 /* --------------------------------------------------------------------------------------------- */
 
 int
-mcview__get_nroff_real_len (mcview_t * view, off_t start, off_t length)
+mcview_get_nroff_real_len (mcview_t * view, off_t start, off_t length)
 {
     mcview_nroff_t *nroff;
     int ret = 0;
diff -urN mc-4.7.1-old/src/viewer/search.c mc-4.7.1-new/src/viewer/search.c
--- mc-4.7.1-old/src/viewer/search.c	2010-02-26 12:01:54.000000000 +0000
+++ mc-4.7.1-new/src/viewer/search.c	2010-03-17 20:19:20.000000000 +0000
@@ -113,7 +113,7 @@
 {
 
     view->search_start = view->search->normal_offset +
-        mcview__get_nroff_real_len (view,
+        mcview_get_nroff_real_len (view,
                                     view->search->start_buffer,
                                     view->search->normal_offset - view->search->start_buffer);
 
@@ -121,7 +121,7 @@
         view->search_start++;
 
     view->search_end = view->search_start + match_len +
-        mcview__get_nroff_real_len (view, view->search_start - 1, match_len);
+        mcview_get_nroff_real_len (view, view->search_start - 1, match_len);
 
     if (view->hex_mode)
     {
@@ -231,7 +231,7 @@
     {
         search_start = (mcview_search_options.backwards) ? -2 : 2;
         search_start = view->search_start + search_start +
-            mcview__get_nroff_real_len (view, view->search_start, 2) * search_start;
+            mcview_get_nroff_real_len (view, view->search_start, 2) * search_start;
     }
     else
     {
diff -urN mc-4.7.1-old/src/widget.c mc-4.7.1-new/src/widget.c
--- mc-4.7.1-old/src/widget.c	2010-02-26 11:34:04.000000000 +0000
+++ mc-4.7.1-new/src/widget.c	2010-03-17 19:16:22.000000000 +0000
@@ -338,16 +338,6 @@
     return b;
 }
 
-const char *
-button_get_text (const WButton *b)
-{
-    if (b->text.hotkey != NULL)
-        return g_strconcat (b->text.start, "&", b->text.hotkey,
-                            b->text.end, (char *) NULL);
-    else
-        return g_strdup (b->text.start);
-}
-
 void
 button_set_text (WButton *b, const char *text)
 {
diff -urN mc-4.7.1-old/src/widget.h mc-4.7.1-new/src/widget.h
--- mc-4.7.1-old/src/widget.h	2010-02-26 11:34:04.000000000 +0000
+++ mc-4.7.1-new/src/widget.h	2010-03-17 20:16:35.000000000 +0000
@@ -209,10 +209,8 @@
 void new_input (WInput *in);
 void stuff (WInput *in, const char *text, int insert_extra_space);
 void input_disable_update (WInput *in);
-void input_set_prompt (WInput *in, int field_len, const char *prompt);
 void input_enable_update (WInput *in);
 void input_set_point (WInput *in, int pos);
-void input_show_cursor (WInput *in);
 void assign_text (WInput *in, const char *text);
 cb_ret_t input_callback (Widget *, widget_msg_t msg, int parm);
 
@@ -225,13 +223,9 @@
 
 /* Buttons */
 /* return copy of button text */
-const char *button_get_text (const WButton *b);
 void button_set_text (WButton *b, const char *text);
 int button_get_len (const WButton *b);
 
-/* Listbox manager */
-WLEntry *listbox_get_data (WListbox *l, int pos);
-
 /* search text int listbox entries */
 int listbox_search_text (WListbox *l, const char *text);
 void listbox_select_entry (WListbox *l, int dest);
