From c661fd15389d3af083dc3bdd20c00ec295f4a494 Mon Sep 17 00:00:00 2001
From: Vit Rosin <vit_r@list.ru>
Date: Thu, 18 Feb 2010 16:23:26 +0000
Subject: [PATCH]  retcheck in lib/skin/ini-file.c

---
 lib/skin/ini-file.c |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/lib/skin/ini-file.c b/lib/skin/ini-file.c
index dc39b6b..e281eec 100644
--- a/lib/skin/ini-file.c
+++ b/lib/skin/ini-file.c
@@ -81,6 +81,8 @@ mc_skin_ini_file_load (mc_skin_t * mc_skin)
     char *file_name, *user_home_dir;
 
     file_name = g_path_get_basename (mc_skin->name);
+    if (file_name == NULL)
+	return FALSE;
 
     if (strcmp (file_name, mc_skin->name) != 0) {
         g_free (file_name);
-- 
1.6.4

