diff --git a/src/key.c b/src/key.c
index c0b20ef..2389c4d 100644
--- a/src/key.c
+++ b/src/key.c
@@ -268,6 +268,10 @@ static key_define_t xterm_key_defines [] = {
     { KEY_M_SHIFT | KEY_M_CTRL | KEY_DOWN,  ESC_STR "[1;6B", MCKEY_NOACTION },
     { KEY_M_SHIFT | KEY_M_CTRL | KEY_RIGHT, ESC_STR "[1;6C", MCKEY_NOACTION },
     { KEY_M_SHIFT | KEY_M_CTRL | KEY_LEFT,  ESC_STR "[1;6D", MCKEY_NOACTION },
+    { KEY_M_SHIFT | KEY_F(1),   ESC_STR "[1;2P",   MCKEY_NOACTION },
+    { KEY_M_SHIFT | KEY_F(2),   ESC_STR "[1;2Q",   MCKEY_NOACTION },
+    { KEY_M_SHIFT | KEY_F(3),   ESC_STR "[1;2R",   MCKEY_NOACTION },
+    { KEY_M_SHIFT | KEY_F(4),   ESC_STR "[1;2S",   MCKEY_NOACTION },
 
     /* putty */
     { KEY_M_SHIFT | KEY_M_CTRL | KEY_UP,    ESC_STR "[[1;6A", MCKEY_NOACTION },
@@ -311,6 +315,10 @@ static key_define_t xterm_key_defines [] = {
     { KEY_M_SHIFT | KEY_M_CTRL | KEY_DOWN,  ESC_STR "[6B", MCKEY_NOACTION },
     { KEY_M_SHIFT | KEY_M_CTRL | KEY_RIGHT, ESC_STR "[6C", MCKEY_NOACTION },
     { KEY_M_SHIFT | KEY_M_CTRL | KEY_LEFT,  ESC_STR "[6D", MCKEY_NOACTION },
+    { KEY_M_SHIFT | KEY_F(1),   ESC_STR "O1;2P",   MCKEY_NOACTION },
+    { KEY_M_SHIFT | KEY_F(2),   ESC_STR "O1;2Q",   MCKEY_NOACTION },
+    { KEY_M_SHIFT | KEY_F(3),   ESC_STR "O1;2R",   MCKEY_NOACTION },
+    { KEY_M_SHIFT | KEY_F(4),   ESC_STR "O1;2S",   MCKEY_NOACTION },
 
     /* gnome-terminal - application mode */
     { KEY_M_CTRL  | KEY_UP,    ESC_STR "O5A",  MCKEY_NOACTION },
