diff --git a/doc/man/mc.1.in b/doc/man/mc.1.in
index aad40fe27..06ea6af41 100644
--- a/doc/man/mc.1.in
+++ b/doc/man/mc.1.in
@@ -1876,6 +1876,11 @@ to
 .BR No .
 This option is disabled by default.
 .PP
+.I Auto cd new terminal.
+If this option is enabled, mc tells its current directory to the terminal.
+Some graphical terminal emulators use this information to open new windows
+or tabs in that directory.
+.PP
 .I Auto save setup.
 If this option is enabled, when you exit Midnight Commander, the
 configurable options of Midnight Commander are saved in the
diff --git a/po/be.po b/po/be.po
index d19484333..c7ed014e5 100644
Binary files a/po/be.po and b/po/be.po differ
diff --git a/po/br.po b/po/br.po
index 2811377a7..4de791b06 100644
Binary files a/po/br.po and b/po/br.po differ
diff --git a/po/hr.po b/po/hr.po
index 30389ce2a..de1f722a0 100644
Binary files a/po/hr.po and b/po/hr.po differ
diff --git a/src/execute.c b/src/execute.c
index 21095a20d..ad60ef96b 100644
--- a/src/execute.c
+++ b/src/execute.c
@@ -373,6 +373,7 @@ do_executev (const char *shell, int flags, char *const argv[])
     {
         update_panels (UP_OPTIMIZE, UP_KEEPSEL);
         update_xterm_title_path ();
+        update_terminal_cwd ();
     }
 
     do_refresh ();
@@ -567,6 +568,7 @@ toggle_subshell (void)
     {
         update_panels (UP_OPTIMIZE, UP_KEEPSEL);
         update_xterm_title_path ();
+        update_terminal_cwd ();
     }
 
     if (was_sigwinch != 0 || tty_got_winch ())
diff --git a/src/filemanager/boxes.c b/src/filemanager/boxes.c
index 2dbc75d65..a6fc3f1d5 100644
--- a/src/filemanager/boxes.c
+++ b/src/filemanager/boxes.c
@@ -580,11 +580,11 @@ configure_box (void)
                                     &mc_global.widget.show_all_if_ambiguous, NULL),
                     QUICK_CHECKBOX (N_("Rotating d&ash"), &nice_rotating_dash, NULL),
                     QUICK_CHECKBOX (N_("Cd follows lin&ks"), &mc_global.vfs.cd_symlinks, NULL),
+                    QUICK_CHECKBOX (N_("Auto cd new terminal"), &auto_cd_new_terminal, NULL),  /* w/o hotkey */
                     QUICK_CHECKBOX (N_("Sa&fe delete"), &safe_delete, NULL),
                     QUICK_CHECKBOX (N_("Safe overwrite"), &safe_overwrite, NULL),       /* w/o hotkey */
                     QUICK_CHECKBOX (N_("A&uto save setup"), &auto_save_setup, NULL),
                     QUICK_SEPARATOR (FALSE),
-                    QUICK_SEPARATOR (FALSE),
                 QUICK_STOP_GROUPBOX,
             QUICK_STOP_COLUMNS,
             QUICK_BUTTONS_OK_CANCEL,
diff --git a/src/filemanager/layout.c b/src/filemanager/layout.c
index 11607720d..bb1bd89d8 100644
--- a/src/filemanager/layout.c
+++ b/src/filemanager/layout.c
@@ -913,6 +913,7 @@ setup_panels (void)
     widget_set_visibility (WIDGET (the_bar), mc_global.keybar_visible);
 
     update_xterm_title_path ();
+    update_terminal_cwd ();
 
     /* unlock */
     if (active)
@@ -1567,7 +1568,7 @@ update_xterm_title_path (void)
         g_free (login);
         g_free (path);
 
-        fprintf (stdout, "\33]0;%s\7", str_term_form (p));
+        fprintf (stdout, "\33]0;%s\33\\", str_term_form (p));
         g_free (p);
 
         if (!mc_global.tty.alternate_plus_minus)
@@ -1577,3 +1578,34 @@ update_xterm_title_path (void)
 }
 
 /* --------------------------------------------------------------------------------------------- */
+
+/** Tell the current directory to the terminal so it can open new tabs there */
+void
+update_terminal_cwd (void)
+{
+    if (mc_global.tty.xterm_flag && auto_cd_new_terminal && vfs_current_is_local ())
+    {
+        const gchar *host;
+        char *path;
+        int pathlen, i;
+        unsigned char c;
+
+        host = g_get_host_name ();
+        path = vfs_path_to_str_flags (current_panel->cwd_vpath, 0, VPF_NONE);
+        pathlen = strlen (path);
+
+        fprintf (stdout, "\33]7;file://%s", host);
+        for (i = 0; i < pathlen; i++) {
+            c = path[i];
+            if ((c >= '0' && c <= '9') || ((c & ~0x20) >= 'A' && (c & ~0x20) <= 'Z') || index("/_.-()~", c) != NULL)
+                fprintf(stdout, "%c", c);
+            else
+                fprintf(stdout, "%%%02X", c);
+        }
+        fprintf (stdout, "\33\\");
+        (void) fflush (stdout);
+        g_free (path);
+    }
+}
+
+/* --------------------------------------------------------------------------------------------- */
diff --git a/src/filemanager/layout.h b/src/filemanager/layout.h
index 2566cfa32..685f2c371 100644
--- a/src/filemanager/layout.h
+++ b/src/filemanager/layout.h
@@ -90,6 +90,7 @@ int load_prompt (int fd, void *unused);
 #endif
 
 void update_xterm_title_path (void);
+void update_terminal_cwd (void);
 
 void title_path_prepare (char **path, char **login);
 
diff --git a/src/filemanager/panel.c b/src/filemanager/panel.c
index b1174c0bf..f14fd38d7 100644
--- a/src/filemanager/panel.c
+++ b/src/filemanager/panel.c
@@ -3436,6 +3436,7 @@ panel_do_cd_int (WPanel * panel, const vfs_path_t * new_dir_vpath, enum cd_enum
     load_hint (FALSE);
     panel->dirty = TRUE;
     update_xterm_title_path ();
+    update_terminal_cwd ();
 
     vfs_path_free (olddir_vpath, TRUE);
 
@@ -3823,6 +3824,7 @@ panel_callback (Widget * w, Widget * sender, widget_msg_t msg, int parm, void *d
             subshell_chdir (panel->cwd_vpath);
 
         update_xterm_title_path ();
+        update_terminal_cwd ();
         select_item (panel);
 
         bb = buttonbar_find (h);
diff --git a/src/setup.c b/src/setup.c
index 9d6e15348..b188de3e8 100644
--- a/src/setup.c
+++ b/src/setup.c
@@ -112,6 +112,9 @@ gboolean confirm_view_dir = FALSE;
 /* Ask file name before start the editor */
 gboolean editor_ask_filename_before_edit = FALSE;
 
+/* Tell cwd to the terminal so it can open new tabs there */
+gboolean auto_cd_new_terminal = TRUE;
+
 panel_view_mode_t startup_left_mode;
 panel_view_mode_t startup_right_mode;
 
@@ -366,6 +369,7 @@ static const struct
     { "mcview_remember_file_position", &mcview_remember_file_position },
     { "auto_fill_mkdir_name", &auto_fill_mkdir_name },
     { "copymove_persistent_attr", &copymove_persistent_attr },
+    { "auto_cd_new_terminal", &auto_cd_new_terminal },
     { NULL, NULL }
 };
 
diff --git a/src/setup.h b/src/setup.h
index b6e675d75..bbee32669 100644
--- a/src/setup.h
+++ b/src/setup.h
@@ -100,6 +100,7 @@ extern gboolean use_file_to_check_type;
 #endif
 extern gboolean file_op_compute_totals;
 extern gboolean editor_ask_filename_before_edit;
+extern gboolean auto_cd_new_terminal;
 
 extern panels_options_t panels_options;
 
