diff -Naur mc-4.7.0-pre1~/edit/edit.c mc-4.7.0-pre1/edit/edit.c
--- mc-4.7.0-pre1~/edit/edit.c	2009-07-31 19:36:25.000000000 +0300
+++ mc-4.7.0-pre1/edit/edit.c	2009-08-10 16:42:47.000000000 +0300
@@ -1541,7 +1541,7 @@
 
 void edit_update_curs_col (WEdit * edit)
 {
-    edit->curs_col = edit_move_forward3(edit, edit_bol(edit, edit->curs1), 0, edit->curs1) + option_line_state_width;
+    edit->curs_col = edit_move_forward3(edit, edit_bol(edit, edit->curs1), 0, edit->curs1);
 }
 
 int
diff -Naur mc-4.7.0-pre1~/edit/editcmd_dialogs.c mc-4.7.0-pre1/edit/editcmd_dialogs.c
--- mc-4.7.0-pre1~/edit/editcmd_dialogs.c	2009-07-31 19:36:25.000000000 +0300
+++ mc-4.7.0-pre1/edit/editcmd_dialogs.c	2009-08-10 13:46:52.000000000 +0300
@@ -267,7 +267,7 @@
     /* calculate the dialog metrics */
     compl_dlg_h = num_compl + 2;
     compl_dlg_w = max_len + 4;
-    start_x = edit->curs_col + edit->start_col - (compl_dlg_w / 2);
+    start_x = edit->curs_col + edit->start_col - (compl_dlg_w / 2) + EDIT_TEXT_HORIZONTAL_OFFSET + option_line_state_width;
     start_y = edit->curs_row + EDIT_TEXT_VERTICAL_OFFSET + 1;
 
     if (start_x < 0)
@@ -335,7 +335,7 @@
     /* calculate the dialog metrics */
     def_dlg_h = num_lines + 2;
     def_dlg_w = max_len + 4;
-    start_x = edit->curs_col + edit->start_col - (def_dlg_w / 2);
+    start_x = edit->curs_col + edit->start_col - (def_dlg_w / 2) + EDIT_TEXT_HORIZONTAL_OFFSET + option_line_state_width;
     start_y = edit->curs_row + EDIT_TEXT_VERTICAL_OFFSET + 1;
 
     if (start_x < 0)
diff -Naur mc-4.7.0-pre1~/edit/editdraw.c mc-4.7.0-pre1/edit/editdraw.c
--- mc-4.7.0-pre1~/edit/editdraw.c	2009-07-31 19:36:25.000000000 +0300
+++ mc-4.7.0-pre1/edit/editdraw.c	2009-08-10 16:52:45.000000000 +0300
@@ -214,6 +214,9 @@
     if (edit->num_widget_lines <= 0 || edit->num_widget_columns <= 0)
 	return;
 
+    edit->num_widget_columns -= EDIT_TEXT_HORIZONTAL_OFFSET + option_line_state_width;
+    edit->num_widget_lines -= EDIT_TEXT_VERTICAL_OFFSET - 1;
+
     r_extreme = EDIT_RIGHT_EXTREME;
     l_extreme = EDIT_LEFT_EXTREME;
     b_extreme = EDIT_BOTTOM_EXTREME;
@@ -250,6 +253,9 @@
     if (outby > 0)
 	edit_scroll_upward (edit, outby);
     edit_update_curs_row (edit);
+
+    edit->num_widget_lines += EDIT_TEXT_VERTICAL_OFFSET - 1;
+    edit->num_widget_columns += EDIT_TEXT_HORIZONTAL_OFFSET + option_line_state_width;
 }
 
 #define set_color(font)    attrset (font)
@@ -274,14 +280,14 @@
 {
     struct line_s *p;
 
-    int x = start_col_real + EDIT_TEXT_HORIZONTAL_OFFSET;
+    int x = start_col_real;
     int x1 = start_col + EDIT_TEXT_HORIZONTAL_OFFSET + option_line_state_width;
     int y = row + EDIT_TEXT_VERTICAL_OFFSET;
     int cols_to_skip = abs (x);
     unsigned char str[6 + 1];
     set_color (EDITOR_NORMAL_COLOR);
     edit_move (x1, y);
-    hline (' ', end_col + 1 - EDIT_TEXT_HORIZONTAL_OFFSET - x1);
+    hline (' ', end_col + 1 - start_col);
 
     if (option_line_state) {
         int i;
@@ -376,6 +382,11 @@
     unsigned int cur_line = 0;
     char line_stat[LINE_STATE_WIDTH + 1];
 
+    if (row > edit->num_widget_lines - EDIT_TEXT_VERTICAL_OFFSET) {
+	return;
+    }
+
+    end_col -= EDIT_TEXT_HORIZONTAL_OFFSET + option_line_state_width;
     edit_get_syntax_color (edit, b - 1, &color);
     q = edit_move_forward3 (edit, b, start_col - edit->start_col, 0);
     start_col_real = (col =
diff -Naur mc-4.7.0-pre1~/edit/editwidget.c mc-4.7.0-pre1/edit/editwidget.c
--- mc-4.7.0-pre1~/edit/editwidget.c	2009-07-28 09:43:41.000000000 +0300
+++ mc-4.7.0-pre1/edit/editwidget.c	2009-08-10 13:58:35.000000000 +0300
@@ -90,7 +90,7 @@
     if (event->type & (GPM_DOWN | GPM_UP))
 	edit_push_key_press (edit);
 
-    edit->prev_col = event->x - edit->start_col - 1 - option_line_state_width;
+    edit->prev_col = event->x - edit->start_col - 1;
 
     if (--event->y > (edit->curs_row + 1))
 	edit_move_down (edit, event->y - (edit->curs_row + 1), 0);
@@ -357,7 +357,7 @@
 
     case WIDGET_CURSOR:
 	widget_move (&e->widget, e->curs_row + EDIT_TEXT_VERTICAL_OFFSET,
-		     e->curs_col + e->start_col);
+		     e->curs_col + e->start_col + EDIT_TEXT_HORIZONTAL_OFFSET + option_line_state_width);
 	return MSG_HANDLED;
 
     case WIDGET_DESTROY:
