diff --git a/src/editor/edit.c b/src/editor/edit.c
index 9f083af..54b22b7 100644
--- a/src/editor/edit.c
+++ b/src/editor/edit.c
@@ -2925,8 +2925,8 @@ edit_move_forward3 (WEdit * edit, long current, int cols, long upto)
     for (col = 0, p = current; p < q; p++)
     {
         int c, orig_c;
+#ifdef HAVE_CHARSET
         int utf_ch = 0;
-#ifdef HAVE_CHARSET
         int cw = 1;
 #endif
         if (cols != -10)
@@ -2951,8 +2951,8 @@ edit_move_forward3 (WEdit * edit, long current, int cols, long upto)
             else if (cw > 1 && g_unichar_isprint (utf_ch))
                 col -= cw - 1;
         }
-#endif
         c = convert_to_display_c (c);
+#endif
         if (c == '\t')
             col += TAB_SIZE - col % TAB_SIZE;
         else if (c == '\n')
