diff -Naur mc-4.7.0-pre1~/src/screen.c mc-4.7.0-pre1/src/screen.c
--- mc-4.7.0-pre1~/src/screen.c	2009-07-31 19:36:26.000000000 +0300
+++ mc-4.7.0-pre1/src/screen.c	2009-08-02 20:14:40.000000000 +0300
@@ -793,6 +793,42 @@
 }
 
 static void
+panel_scrollbar (WPanel *p)
+{
+    int line, n = llines(p);
+
+    if (p->count <= n || slow_terminal) {
+	return;
+    }
+
+    line = 2 + n * p->selected / p->count;
+#if 0
+    /* Are we at the top? */
+    if (p->selected) {
+	widget_move (&p->widget, 1, p->widget.cols - 1);
+	addch ('^');
+    }
+    /* Are we at the bottom? */
+    if (p->selected != p->count - 1) {
+	widget_move (&p->widget, n + 2, p->widget.cols - 1);
+	addch ('v');
+    }
+{
+    int i;
+    for (i = 2; i < n + 2; i++) {
+	if (i != line) {
+	    widget_move (&p->widget, i, p->widget.cols - 1);
+	    addch (0xB0);	/* 0xB0, 0xB1, 0xB2, 0xDB */
+	}
+    }
+}
+#endif
+    /* Now draw the nice relative pointer */
+    widget_move (&p->widget, line, p->widget.cols - 1);
+    addch ('*');		/* REVERSE, MARKED_COLOR: 0xF9 */
+}
+
+static void
 show_dir (WPanel *panel)
 {
     int len;
@@ -810,8 +846,10 @@
 	tty_print_alt_char (ACS_RTEE);
     }
 
-    if (panel->active)
+    if (panel->active) {
+	panel_scrollbar (panel);
 	attrset (REVERSE_COLOR);
+    }
 
     widget_move (&panel->widget, 0, 3);
 
