--- m4.include/mc-with-screen-ncurses.m4.orig	2012-11-13 20:13:18.915951008 +0100
+++ m4.include/mc-with-screen-ncurses.m4	2012-11-13 20:13:08.926153642 +0100
@@ -13,7 +13,7 @@
     fi
 
     saved_CPPFLAGS="$CPPFLAGS"
-    CPPFLAGS="$CPPFLAGS $ac_ncurses_inc_path"
+    CPPFLAGS="$CPPFLAGS $ac_ncurses_inc_path $ac_ncurses_lib_path"
 
     dnl Check for the headers
     dnl Both headers should be in the same directory
--- configure.orig	2012-11-13 20:15:15.260255781 +0100
+++ configure	2012-11-13 20:15:05.907112448 +0100
@@ -15822,7 +15822,7 @@
     fi
 
     saved_CPPFLAGS="$CPPFLAGS"
-    CPPFLAGS="$CPPFLAGS $ac_ncurses_inc_path"
+    CPPFLAGS="$CPPFLAGS $ac_ncurses_inc_path $ac_ncurses_lib_path"
 
                 { $as_echo "$as_me:${as_lineno-$LINENO}: checking for ncurses/ncurses.h and ncurses/term.h" >&5
 $as_echo_n "checking for ncurses/ncurses.h and ncurses/term.h... " >&6; }
