diff --git a/src/layout.c b/src/layout.c
index 4ec5162..5a79818 100644
--- a/src/layout.c
+++ b/src/layout.c
@@ -1165,7 +1165,8 @@ void save_panel_dir(int index)
 	char *widget_work_dir = w->cwd;
 
 	g_free(panels [index].last_saved_dir);  /* last path no needed */
-	panels [index].last_saved_dir = g_strdup(widget_work_dir);
+	/* Because path can nonlocal */
+	panels [index].last_saved_dir = vfs_translate_url(widget_work_dir);
     }
 }
 
@@ -1174,6 +1175,7 @@ Widget *restore_into_right_dir_panel(int index, Widget *from_widget)
 {
     Widget *new_widget = 0;
     const char *saved_dir = panels [index].last_saved_dir;
+
     int last_was_panel = (from_widget &&
             get_display_type(index) != view_listing);
 
