From 3273c020051c60e8bfa9e82b83fc4ca6493a7ab5 Mon Sep 17 00:00:00 2001
From: root <root@holmes.(none)>
Date: Thu, 29 Jan 2009 11:43:08 +0000
Subject: [PATCH] Sequences below are for gnome-terminal, which also sets TERM=xterm (ticket# 219)

---
 lib/mc.lib |    4 ++++
 1 files changed, 4 insertions(+), 0 deletions(-)

diff --git a/lib/mc.lib b/lib/mc.lib
index 85d6ea4..f3c75e9 100644
--- a/lib/mc.lib
+++ b/lib/mc.lib
@@ -107,6 +107,10 @@ f18=\e[19;2~
 f19=\e[20;2~
 f20=\e[21;2~
 
+# Sequences below are for gnome-terminal, which also sets TERM=xterm
+f11=\eO1;2P
+f12=\eO1;2Q
+
 # Arrows for both keypad modes (application and normal).
 up=\e[A
 up=\eOA
-- 
1.6.1.1

