diff --git a/edit/editdraw.c b/edit/editdraw.c
index f0951f4..40f6925 100644
--- a/edit/editdraw.c
+++ b/edit/editdraw.c
@@ -282,7 +282,7 @@ edit_draw_this_line (WEdit *edit, long b, long row, long start_col,
 {
     static unsigned int line[MAX_LINE_LEN];
     unsigned int *p = line;
-    long m1 = 0, m2 = 0, q, c1, c2, tws;
+    long m1 = 0, m2 = 0, q, c1, c2;
     int col, start_col_real;
     unsigned int c;
     int color;
@@ -300,6 +300,7 @@ edit_draw_this_line (WEdit *edit, long b, long row, long start_col,
 	eval_marks (edit, &m1, &m2);
 
 	if (row <= edit->total_lines - edit->start_line) {
+	    long tws = 0;
 	    if (use_colors && visible_tws) {
 		tws = edit_eol (edit, b);
 		while (tws > b && ((c = edit_get_byte (edit, tws - 1)) == ' '
diff --git a/edit/editlock.c b/edit/editlock.c
index fb945f2..89374e7 100644
--- a/edit/editlock.c
+++ b/edit/editlock.c
@@ -142,7 +142,9 @@ lock_get_info (const char *lockfname)
     int cnt;
     static char buf[BUF_SIZE];
 
-    if ((cnt = readlink (lockfname, buf, BUF_SIZE - 1)) == -1 || !*buf)
+    cnt = readlink (lockfname, buf, BUF_SIZE - 1);
+
+    if (cnt == -1 || buf ==NULL || !*buf)
 	return NULL;
     buf[cnt] = '\0';
     return buf;
