diff --git a/src/viewer/search.c b/src/viewer/search.c
index f2e2431..4538266 100644
--- a/src/viewer/search.c
+++ b/src/viewer/search.c
@@ -378,7 +378,8 @@ mcview_do_search (WView * view, off_t want_search_start)
     while (search_start > 0 && mcview_may_still_grow (view));
 
     /* After mcview_may_still_grow (view) == FALSE we have remained last chunk. Search there. */
-    if (!found && view->search->error == MC_SEARCH_E_NOTFOUND && !mcview_search_options.backwards
+    if (view->growbuf_in_use && !found && view->search->error == MC_SEARCH_E_NOTFOUND
+        && !mcview_search_options.backwards
         && mcview_find (&vsm, search_start, mcview_get_filesize (view), &match_len))
     {
         mcview_search_show_result (view, match_len);
