diff --git a/src/viewer/display.c b/src/viewer/display.c
index 40e57a6..1f90aec 100644
--- a/src/viewer/display.c
+++ b/src/viewer/display.c
@@ -56,7 +56,7 @@
 
 /*** file scope macro definitions ****************************************************************/
 
-#define BUF_TRUNC_LEN 5         /* The length of the line displays the file size */
+#define BUF_TRUNC_LEN 11        /* The length of the line displays the file size */
 
 /*** file scope type declarations ****************************************************************/
 
@@ -174,7 +174,7 @@ mcview_display_status (WView * view)
         {
             char buffer[BUF_TRUNC_LEN + 1];
 
-            size_trunc_len (buffer, BUF_TRUNC_LEN, mcview_get_filesize (view), 0,
+            size_trunc_len (buffer, sizeof (buffer) - 1, mcview_get_filesize (view), 0,
                             panels_options.kilobyte_si);
             tty_printf ("%9" PRIuMAX "/%s%s %s", (uintmax_t) view->dpy_end,
                         buffer, mcview_may_still_grow (view) ? "+" : " ",
