diff --git a/lib/search/regex.c b/lib/search/regex.c
index 32be35319..cabf2e486 100644
--- a/lib/search/regex.c
+++ b/lib/search/regex.c
@@ -921,7 +921,7 @@ mc_search__run_regex (mc_search_t * lc_mc_search, const void *user_data,
 
                 g_string_append_c (lc_mc_search->regex_buffer, (char) current_chr);
 
-                if ((char) current_chr == '\n' || virtual_pos > end_search)
+                if ((char) current_chr == '\0' || virtual_pos > end_search)
                     break;
             }
         }
@@ -940,7 +940,7 @@ mc_search__run_regex (mc_search_t * lc_mc_search, const void *user_data,
 
                 current_pos++;
 
-                if (current_chr == '\n' || current_pos > end_search)
+                if (current_chr == '\0' || current_pos > end_search)
                     break;
             }
 
