diff --git a/src/filemanager/boxes.c b/src/filemanager/boxes.c
index 1de6f7d..f03ed05 100644
--- a/src/filemanager/boxes.c
+++ b/src/filemanager/boxes.c
@@ -544,15 +544,25 @@ skin_apply (const gchar * skin_override)
 
 /* --------------------------------------------------------------------------------------------- */
 
+static const gchar *
+skin_name_to_label (const gchar *name)
+{
+    if (strcmp (name, "default") == 0)
+        return _("< Default >");
+    return name;
+}
+
+/* --------------------------------------------------------------------------------------------- */
+
 static int
 sel_skin_button (WButton * button, int action)
 {
     int result;
     WListbox *skin_list;
     WDialog *skin_dlg;
-    gchar *skin_name;
+    const gchar *skin_name;
     int lxx, lyy;
-    unsigned int i;
+    unsigned int i, pos;
 
     (void) action;
 
@@ -563,12 +573,20 @@ sel_skin_button (WButton * button, int action)
                     DLG_COMPACT);
 
     skin_list = listbox_new (1, 1, 11, 22, FALSE, NULL);
+    skin_name = "default";
+    listbox_add_item (skin_list, LISTBOX_APPEND_AT_END, 0, skin_name_to_label(skin_name), (void *) skin_name);
+    if (strcmp (skin_name, current_skin_name) == 0)
+        listbox_select_entry (skin_list, 0);
+    pos = 1;
     for (i = 0; i < skin_names->len; i++)
     {
         skin_name = g_ptr_array_index (skin_names, i);
-        listbox_add_item (skin_list, LISTBOX_APPEND_AT_END, 0, skin_name, NULL);
+        if (strcmp (skin_name, "default") == 0)
+            continue;
+        listbox_add_item (skin_list, LISTBOX_APPEND_AT_END, 0, skin_name_to_label(skin_name), (void *) skin_name);
         if (strcmp (skin_name, current_skin_name) == 0)
-            listbox_select_entry (skin_list, i);
+            listbox_select_entry (skin_list, pos);
+        pos++;
     }
 
     add_widget (skin_dlg, skin_list);
@@ -577,13 +595,14 @@ sel_skin_button (WButton * button, int action)
     if (result == B_ENTER)
     {
         Widget *w;
+        gchar *skin_label;
 
-        listbox_get_current (skin_list, &skin_name, NULL);
+        listbox_get_current (skin_list, &skin_label, (void **) &skin_name);
         g_free (current_skin_name);
         current_skin_name = g_strdup (skin_name);
         skin_apply (skin_name);
         w = dlg_find_by_id (WIDGET (button)->owner, skin_name_id);
-        button_set_text (BUTTON (w), str_fit_to_term (skin_name, 20, J_LEFT_FIT));
+        button_set_text (BUTTON (w), str_fit_to_term (skin_label, 20, J_LEFT_FIT));
     }
     dlg_destroy (skin_dlg);
     return 0;
@@ -603,8 +622,8 @@ appearance_box (void)
             QUICK_START_COLUMNS,
                 QUICK_LABEL (N_("Skin:"), NULL),
             QUICK_NEXT_COLUMN,
-                QUICK_BUTTON (str_fit_to_term(current_skin_name, 20, J_LEFT_FIT), B_USER,
-                              sel_skin_button, &skin_name_id),
+                QUICK_BUTTON (str_fit_to_term(skin_name_to_label(current_skin_name), 20, J_LEFT_FIT),
+                              B_USER, sel_skin_button, &skin_name_id),
             QUICK_STOP_COLUMNS,
             QUICK_BUTTONS_OK_CANCEL,
             QUICK_END
