diff --git a/src/viewer/nroff.c b/src/viewer/nroff.c
index eb3a486..7733429 100644
--- a/src/viewer/nroff.c
+++ b/src/viewer/nroff.c
@@ -104,17 +104,12 @@ mcview_display_nroff (mcview_t * view)
     int c;
     int c_prev = 0;
     int c_next = 0;
-    struct hexedit_change_node *curr = view->change_list;
 
     mcview_display_clean (view);
     mcview_display_ruler (view);
 
     /* Find the first displayable changed byte */
     from = view->dpy_start;
-    while (curr && (curr->offset < from))
-    {
-        curr = curr->next;
-    }
 
     tty_setcolor (VIEW_NORMAL_COLOR);
     for (row = 0, col = 0; row < height;)
diff --git a/src/viewer/plain.c b/src/viewer/plain.c
index 8003f3a..11e65d4 100644
--- a/src/viewer/plain.c
+++ b/src/viewer/plain.c
@@ -74,15 +74,12 @@ mcview_display_text (mcview_t * view)
     int cw = 1;
     int c, prev_ch = 0;
     gboolean last_row = TRUE;
-    struct hexedit_change_node *curr = view->change_list;
 
     mcview_display_clean (view);
     mcview_display_ruler (view);
 
     /* Find the first displayable changed byte */
     from = view->dpy_start;
-    while ((curr != NULL) && (curr->offset < from))
-        curr = curr->next;
 
     while (row < height)
     {
