--- src/filemanager/find.c~	2015-03-20 11:06:04.000000000 -0700
+++ src/filemanager/find.c	2015-03-30 08:07:10.554899540 -0700
@@ -1505,13 +1505,15 @@ find_do_edit_file (WButton * button, int
 /* --------------------------------------------------------------------------------------------- */
 
 static void
-setup_gui (void)
+setup_gui (const char *pattern, const char *content)
 {
     size_t i;
     int lines, cols;
     int y;
 
     static gboolean i18n_flag = FALSE;
+    
+    char title[BUF_MEDIUM];
 
     if (!i18n_flag)
     {
@@ -1531,9 +1533,13 @@ setup_gui (void)
     lines = LINES - 4;
     cols = COLS - 16;
 
+    if ( content && *content )
+	snprintf (title, BUF_MEDIUM, _("Find File: '%s' Content: '%s'"), pattern, content);
+    else
+	snprintf (title, BUF_MEDIUM, _("Find File: '%s'"), pattern);
     find_dlg =
         dlg_create (TRUE, 0, 0, lines, cols, dialog_colors, find_callback, NULL, "[Find File]",
-                    _("Find File"), DLG_CENTER);
+                    title, DLG_CENTER);
 
     find_calc_button_locations (find_dlg, TRUE);
 
@@ -1629,7 +1635,7 @@ do_find (const char *start_dir, ssize_t 
     int return_value = 0;
     char *dir_tmp = NULL, *file_tmp = NULL;
 
-    setup_gui ();
+    setup_gui (pattern, content);
 
     /* FIXME: Need to cleanup this, this ought to be passed non-globaly */
     find_pattern = (char *) pattern;
