diff -ur mc-4.8.10.orig/doc/man/mc.1.in mc-4.8.10/doc/man/mc.1.in
--- mc-4.8.10.orig/doc/man/mc.1.in	2013-08-02 17:02:39.000000000 +0200
+++ mc-4.8.10/doc/man/mc.1.in	2013-10-02 00:08:38.150840530 +0200
@@ -1822,12 +1822,12 @@
 of progressbar follows to direction of Copy/Move/Delete operation:
 from left panel to right one and vice versa. Enabled by default.
 .PP
-.I Mkdir autoname
+.I Mkdir autoname.
 When you press F7 to create a new directory, the input line in popup dialog
 will be filled by name of current file or directory in active panel.
 Disabled by default.
 .PP
-.I Preallocate space
+.I Preallocate space.
 Preallocate space for whole target file, if possible, before copy operation.
 Disabled by default.
 .PP
@@ -1857,17 +1857,17 @@
 that you can examine the output of the command.  There are three
 possible settings for this variable:
 .PP
-.IR Never .
+.I Never.
 Means that you do not want to see the output of your command.  If you
 are using the Linux or FreeBSD console or an xterm, you will be able to
 see the output of the command by typing C\-o.
 .PP
-.IR On dumb terminals .
+.I On dumb terminals.
 You will get the pause message on terminals that are not capable of
 showing the output of the last command executed (any terminal that is
 not an xterm or the Linux console).
 .PP
-.IR Always .
+.I Always.
 The program will pause after executing all of your commands.
 .PP
 .B Other options
@@ -1896,7 +1896,7 @@
 internal file viewer\&.
 .\"Internal File Viewer"
 .PP
-.I Ask new file name
+.I Ask new file name.
 If this option is enabled, file name is asked before open new file in editor.
 .PP
 .I Auto menus.
@@ -1965,7 +1965,7 @@
 The rest of the screen area is used for the two directory panels. You
 can specify whether the area is split to the panels in
 .I Vertical
- or
+or
 .I Horizontal
 direction. Panel layout can be changed using Alt\-, (Alt\-comma) shortcut.
 .PP
